In [1]:
import numpy as np
import pandas as pd

In [2]:
sales = pd.read_csv("data/sales_raw.csv")

In [3]:
print(sales.head())

  Transaction_ID        Date Store_ID Product_ID Customer_Segment Channel  \
0      TX0000001  2025-10-17    ST023     PR0036           Repeat   Store   
1      TX0000002  2024-11-19    ST013     PR0018           Repeat   Store   
2      TX0000003  2025-02-09    ST021     PR0030           Repeat   Store   
3      TX0000004  2025-08-20    ST013     PR0024           Repeat   Store   
4      TX0000005  2024-06-05    ST021     PR0011              New   Store   

  Payment_Method  Quantity  Unit_Price  Discount_Pct  Net_Sales  Cost_Amount  \
0            UPI         1        1099            15     934.15       639.01   
1           Card         1         699            40     419.40       412.16   
2            UPI         1         499             0     499.00       295.82   
3           Cash         1        1299            40     779.40       641.97   
4         Wallet         3        1399            10    3777.30      2132.49   

   Gross_Profit Promo_Flag Salesperson_ID  
0        295

In [4]:
print(sales.tail())

      Transaction_ID        Date Store_ID Product_ID Customer_Segment Channel  \
15000      TX0000251  2024-02-29    ST018     PR0030              New   Store   
15001      TX0002501  2024-05-20    ST011     PR0060          Loyalty   Store   
15002      TX0005001  2024-09-28    ST009     PR0050              New  Online   
15003      TX0009001  2025-03-11    ST008     PR0047           Repeat   Store   
15004      TX0012001  2025-11-11    ST002     PR0031           Repeat  Online   

      Payment_Method  Quantity  Unit_Price  Discount_Pct  Net_Sales  \
15000           Cash         2         499             5     948.10   
15001            UPI         1         699             0     699.00   
15002            UPI         1        1099             5    1044.05   
15003         Wallet         2         799            30    1118.60   
15004            UPI         2         599            30     838.60   

       Cost_Amount  Gross_Profit Promo_Flag Salesperson_ID  
15000       591.64       

In [5]:
print(sales.shape)

(15005, 15)


In [6]:
print(sales.columns)

Index(['Transaction_ID', 'Date', 'Store_ID', 'Product_ID', 'Customer_Segment',
       'Channel', 'Payment_Method', 'Quantity', 'Unit_Price', 'Discount_Pct',
       'Net_Sales', 'Cost_Amount', 'Gross_Profit', 'Promo_Flag',
       'Salesperson_ID'],
      dtype='str')


In [7]:
print(sales.dtypes)

Transaction_ID          str
Date                    str
Store_ID                str
Product_ID              str
Customer_Segment        str
Channel                 str
Payment_Method          str
Quantity              int64
Unit_Price            int64
Discount_Pct          int64
Net_Sales           float64
Cost_Amount         float64
Gross_Profit        float64
Promo_Flag              str
Salesperson_ID          str
dtype: object


In [8]:
# 1. missing values

sales.isnull().sum()

Transaction_ID       0
Date                 0
Store_ID             0
Product_ID           0
Customer_Segment    23
Channel              0
Payment_Method       0
Quantity             0
Unit_Price           0
Discount_Pct         0
Net_Sales            0
Cost_Amount          0
Gross_Profit         0
Promo_Flag           0
Salesperson_ID       0
dtype: int64

In [9]:
# 2. Complete duplicated rows
sales.duplicated().sum()

np.int64(5)

In [10]:
# 3. Duplicate Transaction IDs
sales["Transaction_ID"].duplicated().sum()

np.int64(5)

In [11]:
# 4. Customer segment values
sales['Customer_Segment'].value_counts(dropna=False)

Customer_Segment
Repeat     6713
New        5356
Loyalty    2913
NaN          23
Name: count, dtype: int64

In [12]:
# 5. Channel values
sales['Channel'].value_counts(dropna=False)

Channel
Store      13176
Online      1818
On-line       11
Name: count, dtype: int64

In [13]:
# 6. Payment method values
sales['Payment_Method'].value_counts(dropna=False)

Payment_Method
UPI       6286
Card      4787
Cash      3023
Wallet     894
upi         15
Name: count, dtype: int64

In [14]:
# understand the numerical columns
sales[[
    'Quantity',
    'Unit_Price',
    'Discount_Pct',
    'Net_Sales',
    'Cost_Amount',
    'Gross_Profit'
]].describe()

,Quantity,Unit_Price,Discount_Pct,Net_Sales,Cost_Amount,Gross_Profit
count,15005.000000,15005.000000,15005.000000,15005.000000,15005.000000,15005.000000
mean,1.882839,924.418194,15.219593,1475.679404,977.952180,497.727224
std,1.061242,404.777315,11.019878,1137.862246,754.655911,464.762634
min,1.000000,299.000000,0.000000,134.550000,140.550000,-731.880000
25%,1.000000,599.000000,5.000000,679.150000,461.800000,193.530000
50%,2.000000,899.000000,15.000000,1119.300000,725.710000,356.370000
75%,2.000000,1299.000000,25.000000,1948.500000,1283.940000,641.590000
max,5.000000,1599.000000,55.000000,7995.000000,4992.150000,4270.650000


In [15]:
# 1. Find duplicates
duplicate_rows = sales[
    sales.duplicated(subset=['Transaction_ID'],keep=False)
]

duplicate_rows.sort_values('Transaction_ID')

,Transaction_ID,Date,Store_ID,Product_ID,Customer_Segment,Channel,Payment_Method,Quantity,Unit_Price,Discount_Pct,Net_Sales,Cost_Amount,Gross_Profit,Promo_Flag,Salesperson_ID
250,TX0000251,2024-02-29,ST018,PR0030,New,Store,Cash,2,499,5,948.10,591.64,356.46,N,SP028
15000,TX0000251,2024-02-29,ST018,PR0030,New,Store,Cash,2,499,5,948.10,591.64,356.46,N,SP028
2500,TX0002501,2024-05-20,ST011,PR0060,Loyalty,Store,UPI,1,699,0,699.00,379.89,319.11,N,SP043
15001,TX0002501,2024-05-20,ST011,PR0060,Loyalty,Store,UPI,1,699,0,699.00,379.89,319.11,N,SP043
5000,TX0005001,2024-09-28,ST009,PR0050,New,Online,UPI,1,1099,5,1044.05,593.75,450.30,N,SP065
15002,TX0005001,2024-09-28,ST009,PR0050,New,Online,UPI,1,1099,5,1044.05,593.75,450.30,N,SP065
9000,TX0009001,2025-03-11,ST008,PR0047,Repeat,Store,Wallet,2,799,30,1118.60,770.58,348.02,Y,SP032
15003,TX0009001,2025-03-11,ST008,PR0047,Repeat,Store,Wallet,2,799,30,1118.60,770.58,348.02,Y,SP032
12000,TX0012001,2025-11-11,ST002,PR0031,Repeat,Online,UPI,2,599,30,838.60,763.80,74.80,Y,SP029
15004,TX0012001,2025-11-11,ST002,PR0031,Repeat,Online,UPI,2,599,30,838.60,763.80,74.80,Y,SP029


In [16]:
# 2. Missing Customer Segment rows
sales[sales['Customer_Segment'].isnull()]

,Transaction_ID,Date,Store_ID,Product_ID,Customer_Segment,Channel,Payment_Method,Quantity,Unit_Price,Discount_Pct,Net_Sales,Cost_Amount,Gross_Profit,Promo_Flag,Salesperson_ID
630,TX0000631,2024-04-11,ST011,PR0032,NaN,Store,Card,3,699,15,1782.45,1301.28,481.17,N,SP119
1261,TX0001262,2024-10-11,ST023,PR0039,NaN,Store,UPI,4,1399,30,3917.20,3635.04,282.16,Y,SP088
1892,TX0001893,2024-03-26,ST012,PR0051,NaN,Store,Card,1,1199,5,1139.05,649.19,489.86,N,SP003
2523,TX0002524,2025-05-06,ST008,PR0037,NaN,Store,Card,1,1199,30,839.30,770.32,68.98,Y,SP013
3154,TX0003155,2024-11-22,ST005,PR0045,NaN,Store,UPI,2,599,25,898.50,572.10,326.40,Y,SP008
3785,TX0003786,2025-08-07,ST001,PR0004,NaN,Store,Card,2,699,0,1398.00,711.74,686.26,N,SP093
4416,TX0004417,2025-08-15,ST015,PR0035,NaN,Store,Wallet,2,999,5,1898.10,1085.46,812.64,N,SP036
5047,TX0005048,2025-03-06,ST009,PR0002,NaN,Store,UPI,1,499,5,474.05,232.29,241.76,N,SP026
5678,TX0005679,2024-01-08,ST009,PR0052,NaN,Store,UPI,1,1299,10,1169.10,657.41,511.69,N,SP060
6309,TX0006310,2024-06-27,ST019,PR0025,NaN,Store,Card,1,1399,0,1399.00,938.15,460.85,N,SP084


In [17]:
# 3. Investigate of 55% discount
sales[sales['Discount_Pct']==55]

,Transaction_ID,Date,Store_ID,Product_ID,Customer_Segment,Channel,Payment_Method,Quantity,Unit_Price,Discount_Pct,Net_Sales,Cost_Amount,Gross_Profit,Promo_Flag,Salesperson_ID
1102,TX0001103,2024-04-30,ST017,PR0042,New,Store,Card,1,299,55,134.55,140.55,-6.00,Y,SP016
2205,TX0002206,2024-05-29,ST007,PR0055,Loyalty,Store,UPI,3,1599,55,2158.65,2890.53,-731.88,Y,SP058
3308,TX0003309,2025-06-11,ST014,PR0043,Loyalty,Store,UPI,1,399,55,179.55,203.54,-23.99,Y,SP037
4411,TX0004412,2025-02-14,ST005,PR0007,Repeat,Online,UPI,2,999,55,899.10,1311.24,-412.14,Y,SP026
5514,TX0005515,2025-06-12,ST003,PR0013,New,Store,UPI,1,1599,55,719.55,744.87,-25.32,Y,SP001
6617,TX0006618,2024-07-24,ST014,PR0060,New,Store,UPI,1,699,55,314.55,379.89,-65.34,Y,SP052
7720,TX0007721,2024-01-06,ST003,PR0027,Loyalty,Store,Card,1,1599,55,719.55,768.17,-48.62,Y,SP068
8823,TX0008824,2024-03-28,ST020,PR0025,Loyalty,Store,Card,1,1399,55,629.55,938.15,-308.60,Y,SP057
9926,TX0009927,2024-12-18,ST024,PR0009,Repeat,Store,UPI,2,1199,55,1079.10,1325.66,-246.56,Y,SP055
11029,TX0011030,2024-04-19,ST005,PR0038,Loyalty,Store,Card,1,1299,55,584.55,774.30,-189.75,Y,SP035


In [18]:
sales[sales["Discount_Pct"] == 55][
    ["Transaction_ID", "Date", "Store_ID", "Product_ID",
     "Quantity", "Unit_Price", "Discount_Pct",
     "Net_Sales", "Cost_Amount", "Gross_Profit"]
]

,Transaction_ID,Date,Store_ID,Product_ID,Quantity,Unit_Price,Discount_Pct,Net_Sales,Cost_Amount,Gross_Profit
1102,TX0001103,2024-04-30,ST017,PR0042,1,299,55,134.55,140.55,-6.00
2205,TX0002206,2024-05-29,ST007,PR0055,3,1599,55,2158.65,2890.53,-731.88
3308,TX0003309,2025-06-11,ST014,PR0043,1,399,55,179.55,203.54,-23.99
4411,TX0004412,2025-02-14,ST005,PR0007,2,999,55,899.10,1311.24,-412.14
5514,TX0005515,2025-06-12,ST003,PR0013,1,1599,55,719.55,744.87,-25.32
6617,TX0006618,2024-07-24,ST014,PR0060,1,699,55,314.55,379.89,-65.34
7720,TX0007721,2024-01-06,ST003,PR0027,1,1599,55,719.55,768.17,-48.62
8823,TX0008824,2024-03-28,ST020,PR0025,1,1399,55,629.55,938.15,-308.60
9926,TX0009927,2024-12-18,ST024,PR0009,2,1199,55,1079.10,1325.66,-246.56
11029,TX0011030,2024-04-19,ST005,PR0038,1,1299,55,584.55,774.30,-189.75


In [19]:
# 4. Negative-profit transactions
sales[sales['Gross_Profit']<0]

,Transaction_ID,Date,Store_ID,Product_ID,Customer_Segment,Channel,Payment_Method,Quantity,Unit_Price,Discount_Pct,Net_Sales,Cost_Amount,Gross_Profit,Promo_Flag,Salesperson_ID
69,TX0000070,2024-10-01,ST011,PR0055,New,Store,UPI,1,1599,40,959.4,963.51,-4.11,Y,SP073
129,TX0000130,2024-01-03,ST009,PR0055,Repeat,Store,Cash,5,1599,40,4797.0,4817.55,-20.55,Y,SP111
192,TX0000193,2025-03-21,ST009,PR0041,New,Store,Card,5,1599,40,4797.0,4916.95,-119.95,Y,SP103
287,TX0000288,2024-11-30,ST001,PR0041,Repeat,Store,Wallet,1,1599,40,959.4,983.39,-23.99,Y,SP056
297,TX0000298,2025-12-12,ST007,PR0031,Loyalty,Store,UPI,1,599,40,359.4,381.90,-22.50,Y,SP063
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14597,TX0014598,2024-10-29,ST013,PR0055,Repeat,Store,UPI,1,1599,40,959.4,963.51,-4.11,Y,SP048
14650,TX0014651,2025-07-19,ST018,PR0025,Repeat,Store,Card,4,1399,40,3357.6,3752.60,-395.00,Y,SP061
14662,TX0014663,2025-06-07,ST004,PR0037,Repeat,Store,Card,1,1199,40,719.4,770.32,-50.92,Y,SP063
14777,TX0014778,2024-09-10,ST006,PR0058,Repeat,Store,Card,1,499,40,299.4,309.58,-10.18,Y,SP020


In [20]:
# count of less than 0 gross profit
(sales['Gross_Profit']<0).sum()

np.int64(286)

In [21]:
# check the relationship with discount
sales[sales["Gross_Profit"] < 0]["Discount_Pct"].value_counts().sort_index()

Discount_Pct
40    273
55     13
Name: count, dtype: int64

In [22]:
sales_clean = sales.copy()

In [23]:
# Date datatype
sales_clean['Date'] = pd.to_datetime(sales_clean['Date'])

In [24]:
sales_clean['Date'].dtype

dtype('<M8[us]')

In [25]:
sales_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 15005 entries, 0 to 15004
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction_ID    15005 non-null  str           
 1   Date              15005 non-null  datetime64[us]
 2   Store_ID          15005 non-null  str           
 3   Product_ID        15005 non-null  str           
 4   Customer_Segment  14982 non-null  str           
 5   Channel           15005 non-null  str           
 6   Payment_Method    15005 non-null  str           
 7   Quantity          15005 non-null  int64         
 8   Unit_Price        15005 non-null  int64         
 9   Discount_Pct      15005 non-null  int64         
 10  Net_Sales         15005 non-null  float64       
 11  Cost_Amount       15005 non-null  float64       
 12  Gross_Profit      15005 non-null  float64       
 13  Promo_Flag        15005 non-null  str           
 14  Salesperson_ID    15005 non-null 

### Remove confirmed duplicates

In [26]:
sales_clean.shape

(15005, 15)

In [27]:
sales_clean = sales_clean.drop_duplicates()

In [28]:
sales_clean.shape

(15000, 15)

In [29]:
# Verify the duplicate values are deleted or not
sales_clean.duplicated().sum()

np.int64(0)

### Standardize `Channel`

Currently:

Store      13176<br>
Online      1818<br>
On-line       11<br>

Clean:

In [30]:
sales_clean['Channel'] = sales_clean['Channel'].replace({'On-line':'Online'})

In [31]:
sales_clean['Channel'].value_counts()

Channel
Store     13173
Online     1827
Name: count, dtype: int64

### Standardize Payment Method

In [32]:
sales_clean['Payment_Method'] = sales_clean['Payment_Method'].replace({'upi':'UPI'})

In [33]:
sales_clean['Payment_Method'].value_counts()

Payment_Method
UPI       6298
Card      4787
Cash      3022
Wallet     893
Name: count, dtype: int64

### Missing Customer Segment

In [34]:
sales_clean['Customer_Segment'] = (sales_clean['Customer_Segment'].fillna('Unknown'))

In [35]:
sales_clean['Customer_Segment'].value_counts()

Customer_Segment
Repeat     6711
New        5354
Loyalty    2912
Unknown      23
Name: count, dtype: int64

In [36]:
# Calculate Gross Sales

# Gross Sales= Quantity * Unit Price

sales_clean['Gross_Sales_Calc'] = (sales_clean['Quantity']*sales_clean['Unit_Price'])

In [37]:
# Calculate NetSales

# Net Sales = Gross Sales * (1-Discount%/100)

sales_clean['Net_Sales_Calc'] = (sales_clean['Gross_Sales_Calc']*(1-sales_clean['Discount_Pct']/100)).round(2)

In [38]:
# compare with existing net sales
sales_clean["Net_Sales_Match"] = (
    sales_clean["Net_Sales"].round(2) ==
    sales_clean["Net_Sales_Calc"].round(2)
)

In [39]:
sales_clean["Net_Sales_Match"].value_counts()

Net_Sales_Match
True    15000
Name: count, dtype: int64

In [40]:
# Validate Gross Profit

# Gross Profit = Net Sales - Cost Amount

sales_clean["Gross_Profit_Calc"] = (
    sales_clean["Net_Sales"] - sales_clean["Cost_Amount"]
).round(2)

sales_clean["Gross_Profit_Match"] = (
    sales_clean["Gross_Profit"].round(2) ==
    sales_clean["Gross_Profit_Calc"]
)



In [41]:
sales_clean["Gross_Profit_Match"].value_counts()

Gross_Profit_Match
True    15000
Name: count, dtype: int64

In [42]:
sales_clean["Gross_Profit_Calc"] = (
    sales_clean["Net_Sales"] - sales_clean["Cost_Amount"]
).round(2)

In [43]:
sales_clean["Gross_Profit_Match"].value_counts()

Gross_Profit_Match
True    15000
Name: count, dtype: int64

In [44]:
sales_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Transaction_ID      15000 non-null  str           
 1   Date                15000 non-null  datetime64[us]
 2   Store_ID            15000 non-null  str           
 3   Product_ID          15000 non-null  str           
 4   Customer_Segment    15000 non-null  str           
 5   Channel             15000 non-null  str           
 6   Payment_Method      15000 non-null  str           
 7   Quantity            15000 non-null  int64         
 8   Unit_Price          15000 non-null  int64         
 9   Discount_Pct        15000 non-null  int64         
 10  Net_Sales           15000 non-null  float64       
 11  Cost_Amount         15000 non-null  float64       
 12  Gross_Profit        15000 non-null  float64       
 13  Promo_Flag          15000 non-null  str           
 14  S

In [45]:
# calculate Discount amount

# Discount Amount = Gross Sales - Net Sales

sales_clean['Discount_Amount'] = (sales_clean['Gross_Sales_Calc'] - sales_clean['Net_Sales']).round(2)


In [46]:
# Gross Margin %

# Gross Margin % = (Gross Profit / Net Sales) × 100

sales_clean['Gross_Margin_Pct'] = (sales_clean['Gross_Profit'] / sales_clean['Net_Sales'] * 100).round(2)


In [47]:
# Year

sales_clean['Year'] = sales_clean['Date'].dt.year

In [48]:
# Month Number

sales_clean['Month'] = sales_clean['Date'].dt.month

In [49]:
# Month Name

sales_clean['Month_Name'] = (sales_clean['Date'].dt.strftime('%b'))

In [50]:
# Year Month

sales_clean['YearMonth'] = (sales_clean['Date'].dt.strftime('%Y-%m'))

In [51]:
# Verify New Columns

sales_clean[[
    "Gross_Sales_Calc",
    "Discount_Amount",
    "Gross_Margin_Pct",
    "Year",
    "Month",
    "Month_Name",
    "YearMonth"
]].head()

,Gross_Sales_Calc,Discount_Amount,Gross_Margin_Pct,Year,Month,Month_Name,YearMonth
0,1099,164.85,31.59,2025,10,Oct,2025-10
1,699,279.60,1.73,2024,11,Nov,2024-11
2,499,0.00,40.72,2025,2,Feb,2025-02
3,1299,519.60,17.63,2025,8,Aug,2025-08
4,4197,419.70,43.54,2024,6,Jun,2024-06


In [52]:
sales_clean.shape

(15000, 26)

### KPI— Total Net Sales 

In [262]:
total_sales = sales_clean['Net_Sales'].sum()
print(f'Total Net Sales: ₹{total_sales:,.2f}')

Total Net Sales: ₹22,137,921.10


### KPI— Total Gross Profit

In [263]:
total_gross_profit = sales_clean['Gross_Profit'].sum()
print(f'Total Gross Profit: ₹{total_gross_profit:,.2f}')

Total Gross Profit: ₹7,466,848.30


### KPI— Gross Margin %

In [55]:
# Gross Margin % = (Total Gross Profit ÷ Total Net Sales) × 100

gross_margin_pct = (total_gross_profit/total_sales)*100
print(f'Gross Margin %: {gross_margin_pct:,.2f}%')

Gross Margin %: 33.73%


In [56]:
# validate
gross_margin_pct = (
    sales_clean["Gross_Profit"].sum()
    / sales_clean["Net_Sales"].sum()
) * 100
gross_margin_pct

np.float64(33.728769139031755)

### KPI— Average Transaction Value (ATV) 

In [57]:
# Total Transaction

total_transactions = sales_clean['Transaction_ID'].nunique()
print(f'Total Transactions: {total_transactions:,}')

Total Transactions: 15,000


In [58]:
# ATV = Total Net Sales ÷ Number of Transactions

atv = total_sales/total_transactions
print(f'Average Transaction Value: ₹{atv:,.2f}')

Average Transaction Value: ₹1,475.86


### KPI— Units Per Transaction (UPT) 

In [59]:
# Find total units sold

total_units = sales_clean['Quantity'].sum()
print(f'Total Units Sold: {total_units:,}')

Total Units Sold: 28,244


In [60]:
# UPT = Total Units Sold ÷ Total Transactions

upt = total_units/total_transactions
print(f'Units Per Transaction: {upt:,.2f}')

Units Per Transaction: 1.88


### Monthly Sales Trend + MoM Growth %

#### Monthly Sales

In [61]:
# we created ago sales_clean["YearMonth"]

# aggregate month wise net sales

monthly_sales = (sales_clean.groupby('YearMonth')['Net_Sales'].sum().reset_index())

In [62]:
monthly_sales

,YearMonth,Net_Sales
0,2024-01,880042.70
1,2024-02,916650.20
2,2024-03,856126.20
3,2024-04,958010.45
4,2024-05,987446.20
5,2024-06,952496.10
6,2024-07,927954.05
7,2024-08,957685.05
8,2024-09,895014.00
9,2024-10,881817.95


#### Previous Month Sales

In [264]:
monthly_sales['Previous_Month_Sales'] = (monthly_sales['Net_Sales'].shift(1))

In [265]:
monthly_sales

,YearMonth,Net_Sales,Previous_Month_Sales,MoM_Growth,MoM_Growth_Pct,Previous_Year_Sales,YoY_Growth_Pct,YoY_Growth
0,2024-01,880042.70,NaN,NaN,NaN,NaN,NaN,NaN
1,2024-02,916650.20,880042.70,36607.50,4.16,NaN,NaN,NaN
2,2024-03,856126.20,916650.20,-60524.00,-6.60,NaN,NaN,NaN
3,2024-04,958010.45,856126.20,101884.25,11.90,NaN,NaN,NaN
4,2024-05,987446.20,958010.45,29435.75,3.07,NaN,NaN,NaN
5,2024-06,952496.10,987446.20,-34950.10,-3.54,NaN,NaN,NaN
6,2024-07,927954.05,952496.10,-24542.05,-2.58,NaN,NaN,NaN
7,2024-08,957685.05,927954.05,29731.00,3.20,NaN,NaN,NaN
8,2024-09,895014.00,957685.05,-62671.05,-6.54,NaN,NaN,NaN
9,2024-10,881817.95,895014.00,-13196.05,-1.47,NaN,NaN,NaN


#### MoM Growth %

In [266]:
# MoM Growth % = (Current Month Sales − Previous Month Sales) ÷ Previous Month Sales × 100

monthly_sales['MoM_Growth'] = monthly_sales['Net_Sales']-monthly_sales['Previous_Month_Sales']

monthly_sales['MoM_Growth_Pct'] = (
    (monthly_sales['Net_Sales']- monthly_sales['Previous_Month_Sales'])/monthly_sales['Previous_Month_Sales']*100
).round(2)

In [267]:
monthly_sales[[
    'YearMonth',
    'Net_Sales',
    'Previous_Month_Sales',
    'MoM_Growth',
    'MoM_Growth_Pct'
]]

,YearMonth,Net_Sales,Previous_Month_Sales,MoM_Growth,MoM_Growth_Pct
0,2024-01,880042.70,NaN,NaN,NaN
1,2024-02,916650.20,880042.70,36607.50,4.16
2,2024-03,856126.20,916650.20,-60524.00,-6.60
3,2024-04,958010.45,856126.20,101884.25,11.90
4,2024-05,987446.20,958010.45,29435.75,3.07
5,2024-06,952496.10,987446.20,-34950.10,-3.54
6,2024-07,927954.05,952496.10,-24542.05,-2.58
7,2024-08,957685.05,927954.05,29731.00,3.20
8,2024-09,895014.00,957685.05,-62671.05,-6.54
9,2024-10,881817.95,895014.00,-13196.05,-1.47


### KPI— YoY Sales Growth % 

In [268]:
monthly_sales["Previous_Year_Sales"] = (
    monthly_sales["Net_Sales"].shift(12)
)

#### YoY Growth % 

In [269]:
# YoY Growth % = (Current Year Sales − Previous Year Sales) ÷ Previous Year Sales × 100

monthly_sales['YoY_Growth_Pct'] = (
    (
        monthly_sales["Net_Sales"] - monthly_sales["Previous_Year_Sales"]
    )
    /monthly_sales["Previous_Year_Sales"]*100
).round(2)

In [270]:
monthly_sales['YoY_Growth'] = monthly_sales['Net_Sales']-monthly_sales['Previous_Year_Sales']

In [271]:
monthly_sales[[
    "YearMonth",
    "Net_Sales",
    "Previous_Year_Sales",
    "YoY_Growth",
    "YoY_Growth_Pct"
]].tail(12)

,YearMonth,Net_Sales,Previous_Year_Sales,YoY_Growth,YoY_Growth_Pct
12,2025-01,1001269.65,880042.70,121226.95,13.78
13,2025-02,883701.15,916650.20,-32949.05,-3.59
14,2025-03,1011771.80,856126.20,155645.60,18.18
15,2025-04,917881.95,958010.45,-40128.50,-4.19
16,2025-05,924695.70,987446.20,-62750.50,-6.35
17,2025-06,941871.75,952496.10,-10624.35,-1.12
18,2025-07,1013047.60,927954.05,85093.55,9.17
19,2025-08,1046459.15,957685.05,88774.10,9.27
20,2025-09,856794.85,895014.00,-38219.15,-4.27
21,2025-10,872344.35,881817.95,-9473.60,-1.07


### Store Performance Analysis

In [273]:

stores = pd.read_csv("data/stores.csv")

In [274]:
stores.head(3)

,Store_ID,Store_Name,City,Region,Store_Format,Opening_Date,Selling_Area_SqFt
0,ST001,Delhi Store 1,Delhi,North,High Street,2021-02-01,2850
1,ST002,Delhi Store 2,Delhi,North,Mall,2022-03-01,3200
2,ST003,Lucknow Store 1,Lucknow,North,Mall,2023-04-01,3550


In [275]:
stores.shape

(24, 7)

In [276]:
sales_clean.head(2)

,Transaction_ID,Date,Store_ID,Product_ID,Customer_Segment,Channel,Payment_Method,Quantity,Unit_Price,Discount_Pct,...,Net_Sales_Calc,Net_Sales_Match,Gross_Profit_Calc,Gross_Profit_Match,Discount_Amount,Gross_Margin_Pct,Year,Month,Month_Name,YearMonth
0,TX0000001,2025-10-17,ST023,PR0036,Repeat,Store,UPI,1,1099,15,...,934.15,True,295.14,True,164.85,31.59,2025,10,Oct,2025-10
1,TX0000002,2024-11-19,ST013,PR0018,Repeat,Store,Card,1,699,40,...,419.40,True,7.24,True,279.60,1.73,2024,11,Nov,2024-11


In [277]:
# merge the store file with sales_clean

sales_analysis = sales_clean.merge(
    stores,
    on='Store_ID',
    how='left'
)

In [278]:
sales_analysis.head(2)

,Transaction_ID,Date,Store_ID,Product_ID,Customer_Segment,Channel,Payment_Method,Quantity,Unit_Price,Discount_Pct,...,Year,Month,Month_Name,YearMonth,Store_Name,City,Region,Store_Format,Opening_Date,Selling_Area_SqFt
0,TX0000001,2025-10-17,ST023,PR0036,Repeat,Store,UPI,1,1099,15,...,2025,10,Oct,2025-10,Patna Store 1,Patna,East,Neighborhood,2023-12-01,4950
1,TX0000002,2024-11-19,ST013,PR0018,Repeat,Store,Card,1,699,40,...,2024,11,Nov,2024-11,Bengaluru Store 1,Bengaluru,South,High Street,2023-02-01,4250


In [279]:
sales_clean.shape

(15000, 26)

In [280]:
sales_analysis.shape

(15000, 32)

In [281]:
# check store id is not null in master

sales_analysis[
    sales_analysis["Store_Name"].isnull()
]["Store_ID"].unique()

<ArrowStringArray>
[]
Length: 0, dtype: str

In [282]:
top_stores = (
    sales_analysis
    .groupby(
        ["Store_ID", "Store_Name", "City", "Region"],
        as_index=False
    )
    .agg(
        Net_Sales=("Net_Sales", "sum"),
        Gross_Profit=("Gross_Profit", "sum"),
        Transactions=("Transaction_ID", "nunique"),
        Units_Sold=("Quantity", "sum")
    )
)

In [283]:
top_stores

,Store_ID,Store_Name,City,Region,Net_Sales,Gross_Profit,Transactions,Units_Sold
0,ST001,Delhi Store 1,Delhi,North,931726.85,307867.46,628,1186
1,ST002,Delhi Store 2,Delhi,North,856187.75,284132.04,592,1083
2,ST003,Lucknow Store 1,Lucknow,North,909057.45,303952.32,623,1171
3,ST004,Lucknow Store 2,Lucknow,North,1010681.40,346868.23,662,1265
4,ST005,Jaipur Store 1,Jaipur,North,973104.05,330646.30,648,1252
5,ST006,Jaipur Store 2,Jaipur,North,901147.95,299354.08,613,1185
6,ST007,Mumbai Store 1,Mumbai,West,926636.15,308902.94,626,1193
7,ST008,Mumbai Store 2,Mumbai,West,871206.25,301662.72,614,1149
8,ST009,Pune Store 1,Pune,West,991291.15,331001.60,641,1240
9,ST010,Pune Store 2,Pune,West,914209.20,315049.40,646,1179


In [284]:
# Gross margin percent on store level

top_stores["Gross_Margin_Pct"] = (
    top_stores["Gross_Profit"]
    / top_stores["Net_Sales"]
    * 100
).round(2)

In [285]:
top_stores = (
    top_stores
    .sort_values("Net_Sales", ascending=False)
    .head(10)
)

top_stores

,Store_ID,Store_Name,City,Region,Net_Sales,Gross_Profit,Transactions,Units_Sold,Gross_Margin_Pct
3,ST004,Lucknow Store 2,Lucknow,North,1010681.40,346868.23,662,1265,34.32
14,ST015,Hyderabad Store 1,Hyderabad,South,1002190.55,335292.00,655,1237,33.46
8,ST009,Pune Store 1,Pune,West,991291.15,331001.60,641,1240,33.39
4,ST005,Jaipur Store 1,Jaipur,North,973104.05,330646.30,648,1252,33.98
21,ST022,Bhubaneswar Store 2,Bhubaneswar,East,968567.00,315212.00,645,1243,32.54
15,ST016,Hyderabad Store 2,Hyderabad,South,962046.45,331368.71,657,1223,34.44
19,ST020,Kolkata Store 2,Kolkata,East,958730.80,320237.45,614,1204,33.40
12,ST013,Bengaluru Store 1,Bengaluru,South,941694.80,319935.46,636,1201,33.97
0,ST001,Delhi Store 1,Delhi,North,931726.85,307867.46,628,1186,33.04
22,ST023,Patna Store 1,Patna,East,930345.60,317327.43,619,1187,34.11


### Top 10 Store Analysis

#### Make performance summary of all 24 stores 

In [286]:
store_performance = (
    sales_analysis
    .groupby(
        ["Store_ID", "Store_Name", "City", "Region"],
        as_index=False
    )
    .agg(
        Net_Sales=("Net_Sales", "sum"),
        Gross_Profit=("Gross_Profit", "sum"),
        Transactions=("Transaction_ID", "nunique"),
        Units_Sold=("Quantity", "sum")
    )
)

In [287]:
# Gross Margin % of every store

store_performance['Gross_Margin_Pct'] = (
    store_performance['Gross_Profit']
    /store_performance['Net_Sales'] *100
).round(2)

In [288]:
store_performance.shape

(24, 9)

In [289]:
# Top 10 store

top_10_stores = (
    store_performance.sort_values('Net_Sales', ascending=False)
    .head(10)
)

In [290]:
top_10_stores

,Store_ID,Store_Name,City,Region,Net_Sales,Gross_Profit,Transactions,Units_Sold,Gross_Margin_Pct
3,ST004,Lucknow Store 2,Lucknow,North,1010681.40,346868.23,662,1265,34.32
14,ST015,Hyderabad Store 1,Hyderabad,South,1002190.55,335292.00,655,1237,33.46
8,ST009,Pune Store 1,Pune,West,991291.15,331001.60,641,1240,33.39
4,ST005,Jaipur Store 1,Jaipur,North,973104.05,330646.30,648,1252,33.98
21,ST022,Bhubaneswar Store 2,Bhubaneswar,East,968567.00,315212.00,645,1243,32.54
15,ST016,Hyderabad Store 2,Hyderabad,South,962046.45,331368.71,657,1223,34.44
19,ST020,Kolkata Store 2,Kolkata,East,958730.80,320237.45,614,1204,33.40
12,ST013,Bengaluru Store 1,Bengaluru,South,941694.80,319935.46,636,1201,33.97
0,ST001,Delhi Store 1,Delhi,North,931726.85,307867.46,628,1186,33.04
22,ST023,Patna Store 1,Patna,East,930345.60,317327.43,619,1187,34.11


### Analysis of bottom 10 stores

In [291]:
bottom_10_stores = (
    store_performance.sort_values('Net_Sales', ascending= True)
    .head(10)
)

In [292]:
bottom_10_stores

,Store_ID,Store_Name,City,Region,Net_Sales,Gross_Profit,Transactions,Units_Sold,Gross_Margin_Pct
11,ST012,Ahmedabad Store 2,Ahmedabad,West,793103.80,269540.73,561,1014,33.99
13,ST014,Bengaluru Store 2,Bengaluru,South,839088.50,287978.51,587,1096,34.32
1,ST002,Delhi Store 2,Delhi,North,856187.75,284132.04,592,1083,33.19
7,ST008,Mumbai Store 2,Mumbai,West,871206.25,301662.72,614,1149,34.63
17,ST018,Chennai Store 2,Chennai,South,890984.30,287684.46,607,1150,32.29
10,ST011,Ahmedabad Store 1,Ahmedabad,West,894916.10,314689.95,599,1111,35.16
18,ST019,Kolkata Store 1,Kolkata,East,898420.70,297011.00,620,1147,33.06
5,ST006,Jaipur Store 2,Jaipur,North,901147.95,299354.08,613,1185,33.22
2,ST003,Lucknow Store 1,Lucknow,North,909057.45,303952.32,623,1171,33.44
9,ST010,Pune Store 2,Pune,West,914209.20,315049.40,646,1179,34.46


### Identify reason quantitatively 

#### Add ATV and UPT on Store Level

In [293]:
store_performance['ATV'] = (
    store_performance['Net_Sales']
    /store_performance['Transactions']
).round(2)

store_performance['UPT'] = (
    store_performance['Units_Sold']
    /store_performance['Transactions']
).round(2)

#### Find out the overall company benchmark

In [294]:
company_atv = (
    sales_analysis["Net_Sales"].sum()
    / sales_analysis["Transaction_ID"].nunique()
)

company_upt = (
    sales_analysis["Quantity"].sum()
    / sales_analysis["Transaction_ID"].nunique()
)

company_margin = (
    sales_analysis["Gross_Profit"].sum()
    / sales_analysis["Net_Sales"].sum()
    * 100
)

print(f"Company ATV: ₹{company_atv:,.2f}")
print(f"Company UPT: {company_upt:.2f}")
print(f"Company Gross Margin: {company_margin:.2f}%")

Company ATV: ₹1,475.86
Company UPT: 1.88
Company Gross Margin: 33.73%


In [295]:
# See the Bottom 10 with new KPIs

bottom_10_stores = (
    store_performance
    .sort_values("Net_Sales")
    .head(10)
)

bottom_10_stores[[
    "Store_ID",
    "Store_Name",
    "Net_Sales",
    "Transactions",
    "Units_Sold",
    "ATV",
    "UPT",
    "Gross_Margin_Pct"
]]

,Store_ID,Store_Name,Net_Sales,Transactions,Units_Sold,ATV,UPT,Gross_Margin_Pct
11,ST012,Ahmedabad Store 2,793103.80,561,1014,1413.73,1.81,33.99
13,ST014,Bengaluru Store 2,839088.50,587,1096,1429.45,1.87,34.32
1,ST002,Delhi Store 2,856187.75,592,1083,1446.26,1.83,33.19
7,ST008,Mumbai Store 2,871206.25,614,1149,1418.90,1.87,34.63
17,ST018,Chennai Store 2,890984.30,607,1150,1467.85,1.89,32.29
10,ST011,Ahmedabad Store 1,894916.10,599,1111,1494.02,1.85,35.16
18,ST019,Kolkata Store 1,898420.70,620,1147,1449.07,1.85,33.06
5,ST006,Jaipur Store 2,901147.95,613,1185,1470.06,1.93,33.22
2,ST003,Lucknow Store 1,909057.45,623,1171,1459.16,1.88,33.44
9,ST010,Pune Store 2,914209.20,646,1179,1415.18,1.83,34.46


## Discount vs Profitability

In [296]:
discount_analysis = (
    sales_analysis
    .groupby("Discount_Pct", as_index=False)
    .agg(
        Net_Sales=("Net_Sales", "sum"),
        Gross_Profit=("Gross_Profit", "sum"),
        Transactions=("Transaction_ID", "nunique")
    )
)

discount_analysis["Gross_Margin_Pct"] = (
    discount_analysis["Gross_Profit"]
    / discount_analysis["Net_Sales"]
    * 100
).round(2)

discount_analysis

,Discount_Pct,Net_Sales,Gross_Profit,Transactions,Gross_Margin_Pct
0,0,3834820.00,1676247.50,2224,43.71
1,5,3255922.65,1331153.35,1973,40.88
2,10,4052922.30,1534942.75,2534,37.87
3,15,3384963.50,1146159.51,2304,33.86
4,20,2768080.00,820996.94,1987,29.66
5,25,2289102.00,577917.92,1774,25.25
6,30,1717411.50,337304.75,1391,19.64
7,40,826336.80,44305.12,800,5.36
8,55,8362.35,-2179.54,13,-26.06


### Now convert in management insight of 286 loss transaction

In [297]:
loss_transactions = sales_analysis[
    sales_analysis["Gross_Profit"] < 0
]

loss_summary = {
    "Loss_Transactions": loss_transactions["Transaction_ID"].nunique(),
    "Net_Sales": loss_transactions["Net_Sales"].sum(),
    "Gross_Loss": loss_transactions["Gross_Profit"].sum()
}

loss_summary

{'Loss_Transactions': 286,
 'Net_Sales': np.float64(324541.35),
 'Gross_Loss': np.float64(-22086.47)}

In [298]:
# Loss-making Transaction Rate

loss_transaction_pct = (
    loss_transactions["Transaction_ID"].nunique()
    / sales_analysis["Transaction_ID"].nunique()
    * 100
)

print(f"Loss-making Transaction Rate: {loss_transaction_pct:.2f}%")

Loss-making Transaction Rate: 1.91%


In [299]:
# loss_by_discount

loss_by_discount = (
    loss_transactions
    .groupby("Discount_Pct", as_index=False)
    .agg(
        Loss_Transactions=("Transaction_ID", "nunique"),
        Net_Sales=("Net_Sales", "sum"),
        Gross_Loss=("Gross_Profit", "sum")
    )
)

loss_by_discount

,Discount_Pct,Loss_Transactions,Net_Sales,Gross_Loss
0,40,273,316179.00,-19906.93
1,55,13,8362.35,-2179.54


# Root Cause Analysis

#### Why are some transactions profitable at a 40% discount, while others are loss-making?

In [300]:
# Investogate on Product_ID level

loss_by_product = (
    loss_transactions
    .groupby("Product_ID", as_index=False)
    .agg(
        Loss_Transactions=("Transaction_ID", "nunique"),
        Net_Sales=("Net_Sales", "sum"),
        Gross_Loss=("Gross_Profit", "sum")
    )
    .sort_values("Gross_Loss")
)

loss_by_product.head(10)

,Product_ID,Loss_Transactions,Net_Sales,Gross_Loss
12,PR0025,14,24972.15,-3172.35
20,PR0039,17,36094.20,-2982.48
17,PR0034,17,23194.20,-2863.80
24,PR0054,13,22485.00,-2475.75
3,PR0007,13,14085.90,-1648.98
18,PR0037,16,23020.80,-1629.44
10,PR0021,17,23376.60,-1452.36
9,PR0019,13,13423.20,-851.76
15,PR0031,17,13297.80,-832.50
25,PR0055,12,19427.85,-805.86


### Next — Join Product Master

In [301]:
products = pd.read_csv("data/products.csv")

products.head()

,Product_ID,Product_Name,Category,Subcategory,Brand,MRP,Unit_Cost,Status
0,PR0001,StyleHub Shirts 1,Men,Shirts,StyleHub,399,239.67,Active
1,PR0002,NovaWear Shirts 2,Men,Shirts,NovaWear,499,232.29,Active
2,PR0003,TrendCo Shirts 3,Men,Shirts,TrendCo,599,311.78,Active
3,PR0004,Everyday Shirts 4,Men,Shirts,Everyday,699,355.87,Active
4,PR0005,PrimeFit T-Shirts 1,Men,T-Shirts,PrimeFit,799,497.00,Active


In [302]:
products.columns

Index(['Product_ID', 'Product_Name', 'Category', 'Subcategory', 'Brand', 'MRP',
       'Unit_Cost', 'Status'],
      dtype='str')

In [303]:
products.shape

(60, 8)

In [304]:
# Merge loss_by_product with product master

loss_product_detail = loss_by_product.merge(
    products,
    on="Product_ID",
    how="left"
)

In [305]:
loss_product_detail[[
    "Product_ID",
    "Product_Name",
    "Category",
    "Subcategory",
    "Brand",
    "MRP",
    "Unit_Cost",
    "Loss_Transactions",
    "Net_Sales",
    "Gross_Loss"
]].head(10)

,Product_ID,Product_Name,Category,Subcategory,Brand,MRP,Unit_Cost,Loss_Transactions,Net_Sales,Gross_Loss
0,PR0025,StyleHub T-Shirts 1,Kids,T-Shirts,StyleHub,1399,938.15,14,24972.15,-3172.35
1,PR0039,TrendCo Casual 3,Footwear,Casual,TrendCo,1399,908.76,17,36094.20,-2982.48
2,PR0034,Everyday Sets 2,Kids,Sets,Everyday,899,606.00,17,23194.20,-2863.80
3,PR0054,UrbanEdge Belts 2,Accessories,Belts,UrbanEdge,1499,998.43,13,22485.00,-2475.75
4,PR0007,StyleHub T-Shirts 3,Men,T-Shirts,StyleHub,999,655.62,13,14085.90,-1648.98
5,PR0037,StyleHub Casual 1,Footwear,Casual,StyleHub,1199,770.32,16,23020.80,-1629.44
6,PR0021,TrendCo Dresses 1,Women,Dresses,TrendCo,999,636.64,17,23376.60,-1452.36
7,PR0019,StyleHub Tops 3,Women,Tops,StyleHub,799,509.82,13,13423.20,-851.76
8,PR0031,StyleHub Jeans 3,Kids,Jeans,StyleHub,599,381.90,17,13297.80,-832.50
9,PR0055,StyleHub Belts 3,Accessories,Belts,StyleHub,1599,963.51,12,19427.85,-805.86


In [306]:
# Merge Validation

loss_product_detail[
    loss_product_detail["Product_Name"].isnull()
]["Product_ID"].unique()

<ArrowStringArray>
[]
Length: 0, dtype: str

### Category-wise loss concentration

In [307]:
loss_by_category = (
    loss_product_detail
    .groupby("Category", as_index=False)
    .agg(
        Loss_Transactions=("Loss_Transactions", "sum"),
        Net_Sales=("Net_Sales", "sum"),
        Gross_Loss=("Gross_Loss", "sum")
    )
    .sort_values("Gross_Loss")
)

loss_by_category

,Category,Loss_Transactions,Net_Sales,Gross_Loss
2,Kids,72,76200.30,-7531.52
1,Footwear,53,91853.40,-5647.32
0,Accessories,38,48814.20,-3570.91
3,Men,61,50787.75,-2674.03
4,Women,62,56885.70,-2662.69


### Now Category Overall Performance vs Loss can be compared 

In [308]:
sales_analysis.columns

Index(['Transaction_ID', 'Date', 'Store_ID', 'Product_ID', 'Customer_Segment',
       'Channel', 'Payment_Method', 'Quantity', 'Unit_Price', 'Discount_Pct',
       'Net_Sales', 'Cost_Amount', 'Gross_Profit', 'Promo_Flag',
       'Salesperson_ID', 'Gross_Sales_Calc', 'Net_Sales_Calc',
       'Net_Sales_Match', 'Gross_Profit_Calc', 'Gross_Profit_Match',
       'Discount_Amount', 'Gross_Margin_Pct', 'Year', 'Month', 'Month_Name',
       'YearMonth', 'Store_Name', 'City', 'Region', 'Store_Format',
       'Opening_Date', 'Selling_Area_SqFt'],
      dtype='str')

In [309]:
sales_analysis = sales_analysis.merge(
    products,
    on="Product_ID",
    how="left"
)

In [310]:
sales_analysis.shape

(15000, 39)

In [311]:
# Validate product mapping
sales_analysis[
    sales_analysis["Product_Name"].isnull()
]["Product_ID"].unique()

<ArrowStringArray>
[]
Length: 0, dtype: str

In [312]:
sales_analysis.columns

Index(['Transaction_ID', 'Date', 'Store_ID', 'Product_ID', 'Customer_Segment',
       'Channel', 'Payment_Method', 'Quantity', 'Unit_Price', 'Discount_Pct',
       'Net_Sales', 'Cost_Amount', 'Gross_Profit', 'Promo_Flag',
       'Salesperson_ID', 'Gross_Sales_Calc', 'Net_Sales_Calc',
       'Net_Sales_Match', 'Gross_Profit_Calc', 'Gross_Profit_Match',
       'Discount_Amount', 'Gross_Margin_Pct', 'Year', 'Month', 'Month_Name',
       'YearMonth', 'Store_Name', 'City', 'Region', 'Store_Format',
       'Opening_Date', 'Selling_Area_SqFt', 'Product_Name', 'Category',
       'Subcategory', 'Brand', 'MRP', 'Unit_Cost', 'Status'],
      dtype='str')

In [313]:
category_performance = (
    sales_analysis
    .groupby("Category", as_index=False)
    .agg(
        Net_Sales=("Net_Sales", "sum"),
        Gross_Profit=("Gross_Profit", "sum"),
        Transactions=("Transaction_ID", "nunique"),
        Units_Sold=("Quantity", "sum")
    )
)

category_performance

,Category,Net_Sales,Gross_Profit,Transactions,Units_Sold
0,Accessories,4713462.60,1584538.46,3036,5696
1,Footwear,4349526.45,1390052.15,2996,5525
2,Kids,4354673.00,1366368.64,2975,5693
3,Men,4602179.85,1651496.60,3007,5675
4,Women,4118079.20,1474392.45,2986,5655


In [314]:
sales_analysis.shape

(15000, 39)

In [315]:
# Gross Margin %

category_performance["Gross_Margin_Pct"] = (
    category_performance["Gross_Profit"]
    / category_performance["Net_Sales"]
    * 100
).round(2)

In [316]:
# Merge loss data 

category_analysis = category_performance.merge(
    loss_by_category[
        ["Category", "Loss_Transactions", "Gross_Loss"]
    ],
    on="Category",
    how="left"
)

In [317]:
# loss transaction rate

category_analysis["Loss_Transaction_Rate_Pct"] = (
    category_analysis["Loss_Transactions"]
    / category_analysis["Transactions"]
    * 100
).round(2)

In [318]:
category_analysis.sort_values(
    "Loss_Transaction_Rate_Pct",
    ascending=False
)

,Category,Net_Sales,Gross_Profit,Transactions,Units_Sold,Gross_Margin_Pct,Loss_Transactions,Gross_Loss,Loss_Transaction_Rate_Pct
2,Kids,4354673.00,1366368.64,2975,5693,31.38,72,-7531.52,2.42
4,Women,4118079.20,1474392.45,2986,5655,35.80,62,-2662.69,2.08
3,Men,4602179.85,1651496.60,3007,5675,35.89,61,-2674.03,2.03
1,Footwear,4349526.45,1390052.15,2996,5525,31.96,53,-5647.32,1.77
0,Accessories,4713462.60,1584538.46,3036,5696,33.62,38,-3570.91,1.25


## Subcategory Analysis

In [319]:
subcategory_performance = (
    sales_analysis
    .groupby(
        ["Category", "Subcategory"],
        as_index=False
    )
    .agg(
        Net_Sales=("Net_Sales", "sum"),
        Gross_Profit=("Gross_Profit", "sum"),
        Transactions=("Transaction_ID", "nunique"),
        Units_Sold=("Quantity", "sum")
    )
)

subcategory_performance["Gross_Margin_Pct"] = (
    subcategory_performance["Gross_Profit"]
    / subcategory_performance["Net_Sales"]
    * 100
).round(2)

subcategory_performance.sort_values(
    "Gross_Margin_Pct"
)

,Category,Subcategory,Net_Sales,Gross_Profit,Transactions,Units_Sold,Gross_Margin_Pct
6,Kids,Jeans,834905.30,222341.50,934,1793,26.63
3,Footwear,Casual,2141768.40,574178.22,1003,1884,26.81
7,Kids,Sets,1527162.30,451091.69,1010,1916,29.54
1,Accessories,Belts,1901542.40,564137.97,995,1847,29.67
11,Men,T-Shirts,1546399.15,477164.47,1006,1924,30.86
5,Footwear,Sports,1110205.85,368350.48,1033,1903,33.18
12,Women,Dresses,1808442.30,607076.21,967,1851,33.57
8,Kids,T-Shirts,1992605.40,692935.45,1031,1984,34.78
14,Women,Tops,1196525.95,422001.21,985,1890,35.27
0,Accessories,Bags,1947636.25,704802.54,1044,1989,36.19


In [320]:
# subcategory summary of loss-making transactions

loss_subcategory = (
    loss_transactions
    .merge(
        products[
            ["Product_ID", "Category", "Subcategory"]
        ],
        on="Product_ID",
        how="left"
    )
    .groupby(
        ["Category", "Subcategory"],
        as_index=False
    )
    .agg(
        Loss_Transactions=("Transaction_ID", "nunique"),
        Gross_Loss=("Gross_Profit", "sum")
    )
)

In [321]:
# combine both

subcategory_analysis = subcategory_performance.merge(
    loss_subcategory,
    on=["Category", "Subcategory"],
    how="left"
)

In [322]:
# Set missing loss values to zero

subcategory_analysis[
    ["Loss_Transactions", "Gross_Loss"]
] = subcategory_analysis[
    ["Loss_Transactions", "Gross_Loss"]
].fillna(0)

In [323]:
# Loss Rate

subcategory_analysis["Loss_Transaction_Rate_Pct"] = (
    subcategory_analysis["Loss_Transactions"]
    / subcategory_analysis["Transactions"]
    * 100
).round(2)

In [324]:
# Finally

subcategory_analysis.sort_values(
    "Loss_Transaction_Rate_Pct",
    ascending=False
)


,Category,Subcategory,Net_Sales,Gross_Profit,Transactions,Units_Sold,Gross_Margin_Pct,Loss_Transactions,Gross_Loss,Loss_Transaction_Rate_Pct
11,Men,T-Shirts,1546399.15,477164.47,1006,1924,30.86,45.0,-2341.06,4.47
6,Kids,Jeans,834905.30,222341.50,934,1793,26.63,40.0,-1446.75,4.28
3,Footwear,Casual,2141768.40,574178.22,1003,1884,26.81,34.0,-4801.67,3.39
12,Women,Dresses,1808442.30,607076.21,967,1851,33.57,31.0,-1736.22,3.21
1,Accessories,Belts,1901542.40,564137.97,995,1847,29.67,25.0,-3281.61,2.51
5,Footwear,Sports,1110205.85,368350.48,1033,1903,33.18,19.0,-845.65,1.84
13,Women,Kurtis,1113110.95,445315.03,1034,1914,40.01,18.0,-74.71,1.74
7,Kids,Sets,1527162.30,451091.69,1010,1916,29.54,17.0,-2863.80,1.68
8,Kids,T-Shirts,1992605.40,692935.45,1031,1984,34.78,15.0,-3220.97,1.45
10,Men,Shirts,856139.60,332954.83,990,1831,38.89,14.0,-5.13,1.41


## Product-level root cause 

In [325]:
# Now, we will identify individual products within the high-risk sub-categories.

product_performance = (
    sales_analysis
    .groupby(
        [
            "Product_ID",
            "Product_Name",
            "Category",
            "Subcategory"
        ],
        as_index=False
    )
    .agg(
        Net_Sales=("Net_Sales", "sum"),
        Gross_Profit=("Gross_Profit", "sum"),
        Transactions=("Transaction_ID", "nunique"),
        Units_Sold=("Quantity", "sum")
    )
)

product_performance["Gross_Margin_Pct"] = (
    product_performance["Gross_Profit"]
    / product_performance["Net_Sales"]
    * 100
).round(2)

In [326]:
# Now merged with the Loss Product Summary:

product_analysis = product_performance.merge(
    loss_by_product[
        ["Product_ID", "Loss_Transactions", "Gross_Loss"]
    ],
    on="Product_ID",
    how="left"
)

In [327]:
# For products with no loss, 'NaN' will appear; treat them as zero.

product_analysis[
    ["Loss_Transactions", "Gross_Loss"]
] = product_analysis[
    ["Loss_Transactions", "Gross_Loss"]
].fillna(0)

In [328]:
# Loss rate

product_analysis["Loss_Transaction_Rate_Pct"] = (
    product_analysis["Loss_Transactions"]
    / product_analysis["Transactions"]
    * 100
).round(2)

In [329]:
# highest-risk products

product_analysis.sort_values(
    "Loss_Transaction_Rate_Pct",
    ascending=False
).head(15)

,Product_ID,Product_Name,Category,Subcategory,Net_Sales,Gross_Profit,Transactions,Units_Sold,Gross_Margin_Pct,Loss_Transactions,Gross_Loss,Loss_Transaction_Rate_Pct
20,PR0021,TrendCo Dresses 1,Women,Dresses,354395.25,83186.61,222,426,23.47,17.0,-1452.36,7.66
5,PR0006,UrbanEdge T-Shirts 2,Men,T-Shirts,357487.35,97481.85,251,475,27.27,18.0,-287.28,7.17
38,PR0039,TrendCo Casual 3,Footwear,Casual,539804.15,123592.07,240,458,22.90,17.0,-2982.48,7.08
30,PR0031,StyleHub Jeans 3,Kids,Jeans,237563.40,57688.50,241,471,24.28,17.0,-832.50,7.05
33,PR0034,Everyday Sets 2,Kids,Sets,358925.75,68045.75,242,480,18.96,17.0,-2863.80,7.02
14,PR0015,TrendCo Kurtis 3,Women,Kurtis,155330.70,43937.53,233,463,28.29,16.0,-33.32,6.87
31,PR0032,NovaWear Jeans 4,Kids,Jeans,251010.90,67964.18,235,422,27.08,16.0,-373.36,6.81
40,PR0041,PrimeFit Sports 1,Footwear,Sports,622090.95,169731.55,258,460,27.28,16.0,-791.67,6.20
36,PR0037,StyleHub Casual 1,Footwear,Casual,511433.45,123962.49,263,503,24.24,16.0,-1629.44,6.08
0,PR0001,StyleHub Shirts 1,Men,Shirts,148368.15,43392.69,246,438,29.25,14.0,-5.13,5.69


# Strong Analysis: Break-even Discount 

#### Break-even discount % = (1 − Unit Cost / MRP) × 100

#### Add product cost details 

In [330]:
product_analysis = product_analysis.merge(
    products[
        ["Product_ID", "Brand", "MRP", "Unit_Cost"]
    ],
    on="Product_ID",
    how="left"
)

#### Calculate Break-even Discount 

In [331]:
product_analysis["Break_Even_Discount_Pct"] = (
    (
        1 -
        product_analysis["Unit_Cost"]
        / product_analysis["MRP"]
    ) * 100
).round(2)

In [332]:
product_analysis[[
    "Product_ID",
    "Product_Name",
    "Category",
    "MRP",
    "Unit_Cost",
    "Gross_Margin_Pct",
    "Break_Even_Discount_Pct",
    "Loss_Transaction_Rate_Pct",
    "Gross_Loss"
]].sort_values(
    "Loss_Transaction_Rate_Pct",
    ascending=False
).head(15)

,Product_ID,Product_Name,Category,MRP,Unit_Cost,Gross_Margin_Pct,Break_Even_Discount_Pct,Loss_Transaction_Rate_Pct,Gross_Loss
20,PR0021,TrendCo Dresses 1,Women,999,636.64,23.47,36.27,7.66,-1452.36
5,PR0006,UrbanEdge T-Shirts 2,Men,899,547.38,27.27,39.11,7.17,-287.28
38,PR0039,TrendCo Casual 3,Footwear,1399,908.76,22.90,35.04,7.08,-2982.48
30,PR0031,StyleHub Jeans 3,Kids,599,381.90,24.28,36.24,7.05,-832.50
33,PR0034,Everyday Sets 2,Kids,899,606.00,18.96,32.59,7.02,-2863.80
14,PR0015,TrendCo Kurtis 3,Women,399,240.59,28.29,39.70,6.87,-33.32
31,PR0032,NovaWear Jeans 4,Kids,699,433.76,27.08,37.95,6.81,-373.36
40,PR0041,PrimeFit Sports 1,Footwear,1599,983.39,27.28,38.50,6.20,-791.67
36,PR0037,StyleHub Casual 1,Footwear,1199,770.32,24.24,35.75,6.08,-1629.44
0,PR0001,StyleHub Shirts 1,Men,399,239.67,29.25,39.93,5.69,-5.13


#### Products will now be classified into risk bands:

In [333]:
import numpy as np

product_analysis["Discount_Risk"] = np.where(
    product_analysis["Break_Even_Discount_Pct"] < 40,
    "High Risk",
    "Normal Risk"
)

In [334]:
product_analysis["Discount_Risk"].value_counts()

Discount_Risk
Normal Risk    41
High Risk      19
Name: count, dtype: int64

In [335]:
# High Risk products

high_risk_products = (
    product_analysis[
        product_analysis["Discount_Risk"] == "High Risk"
    ]
    .sort_values("Break_Even_Discount_Pct")
)

high_risk_products[[
    "Product_ID",
    "Product_Name",
    "Category",
    "MRP",
    "Unit_Cost",
    "Break_Even_Discount_Pct",
    "Loss_Transaction_Rate_Pct",
    "Gross_Loss"
]]

,Product_ID,Product_Name,Category,MRP,Unit_Cost,Break_Even_Discount_Pct,Loss_Transaction_Rate_Pct,Gross_Loss
33,PR0034,Everyday Sets 2,Kids,899,606.00,32.59,7.02,-2863.80
24,PR0025,StyleHub T-Shirts 1,Kids,1399,938.15,32.94,5.60,-3172.35
53,PR0054,UrbanEdge Belts 2,Accessories,1499,998.43,33.39,5.31,-2475.75
6,PR0007,StyleHub T-Shirts 3,Men,999,655.62,34.37,5.22,-1648.98
38,PR0039,TrendCo Casual 3,Footwear,1399,908.76,35.04,7.08,-2982.48
28,PR0029,PrimeFit Jeans 1,Kids,399,257.93,35.36,3.02,-240.89
36,PR0037,StyleHub Casual 1,Footwear,1199,770.32,35.75,6.08,-1629.44
18,PR0019,StyleHub Tops 3,Women,799,509.82,36.19,5.18,-851.76
30,PR0031,StyleHub Jeans 3,Kids,599,381.90,36.24,7.05,-832.50
20,PR0021,TrendCo Dresses 1,Women,999,636.64,36.27,7.66,-1452.36


In [336]:
# Just "High Risk" / "Normal Risk" is a bit basic. Create three risk bands using Numpy.

conditions = [
    product_analysis["Break_Even_Discount_Pct"] < 35,
    product_analysis["Break_Even_Discount_Pct"] < 40
]

choices = [
    "Critical Risk",
    "High Risk"
]

product_analysis["Discount_Risk_Band"] = np.select(
    conditions,
    choices,
    default="Normal Risk"
)

In [337]:
product_analysis["Discount_Risk_Band"].value_counts()

Discount_Risk_Band
Normal Risk      41
High Risk        15
Critical Risk     4
Name: count, dtype: int64

In [338]:
product_analysis[
    [
        "Product_ID",
        "Product_Name",
        "Break_Even_Discount_Pct",
        "Discount_Risk_Band"
    ]
].sort_values("Break_Even_Discount_Pct").head(15)

,Product_ID,Product_Name,Break_Even_Discount_Pct,Discount_Risk_Band
33,PR0034,Everyday Sets 2,32.59,Critical Risk
24,PR0025,StyleHub T-Shirts 1,32.94,Critical Risk
53,PR0054,UrbanEdge Belts 2,33.39,Critical Risk
6,PR0007,StyleHub T-Shirts 3,34.37,Critical Risk
38,PR0039,TrendCo Casual 3,35.04,High Risk
28,PR0029,PrimeFit Jeans 1,35.36,High Risk
36,PR0037,StyleHub Casual 1,35.75,High Risk
18,PR0019,StyleHub Tops 3,36.19,High Risk
30,PR0031,StyleHub Jeans 3,36.24,High Risk
20,PR0021,TrendCo Dresses 1,36.27,High Risk


### Now, we will validate the rule against actual transactions.

In [339]:
# First, add the break-even value to the transaction data.

break_even_lookup = product_analysis[
    ["Product_ID", "Break_Even_Discount_Pct"]
]

discount_40 = (
    sales_analysis[sales_analysis["Discount_Pct"] == 40]
    .merge(
        break_even_lookup,
        on="Product_ID",
        how="left"
    )
)

In [340]:
# Now, theoretically, should a loss occur or not?

discount_40["Expected_Loss"] = (
    discount_40["Discount_Pct"]
    >
    discount_40["Break_Even_Discount_Pct"]
)

In [341]:
# Actual transaction loss

discount_40["Actual_Loss"] = (
    discount_40["Gross_Profit"] < 0
)

In [342]:
# Comparision

pd.crosstab(
    discount_40["Expected_Loss"],
    discount_40["Actual_Loss"]
)

Actual_Loss,False,True
Expected_Loss,,
False,527,0
True,0,273


### Management Recommendation

### Recommended discount cap 

In [343]:
product_analysis["Recommended_Max_Discount_Pct"] = (
    np.floor(
        (product_analysis["Break_Even_Discount_Pct"] - 5) / 5
    ) * 5
).clip(lower=0)

In [344]:
# Now, take a look at the important products.

product_analysis[[
    "Product_ID",
    "Product_Name",
    "Category",
    "Break_Even_Discount_Pct",
    "Recommended_Max_Discount_Pct",
    "Discount_Risk_Band"
]].sort_values(
    "Break_Even_Discount_Pct"
).head(20)

,Product_ID,Product_Name,Category,Break_Even_Discount_Pct,Recommended_Max_Discount_Pct,Discount_Risk_Band
33,PR0034,Everyday Sets 2,Kids,32.59,25.0,Critical Risk
24,PR0025,StyleHub T-Shirts 1,Kids,32.94,25.0,Critical Risk
53,PR0054,UrbanEdge Belts 2,Accessories,33.39,25.0,Critical Risk
6,PR0007,StyleHub T-Shirts 3,Men,34.37,25.0,Critical Risk
38,PR0039,TrendCo Casual 3,Footwear,35.04,30.0,High Risk
28,PR0029,PrimeFit Jeans 1,Kids,35.36,30.0,High Risk
36,PR0037,StyleHub Casual 1,Footwear,35.75,30.0,High Risk
18,PR0019,StyleHub Tops 3,Women,36.19,30.0,High Risk
30,PR0031,StyleHub Jeans 3,Kids,36.24,30.0,High Risk
20,PR0021,TrendCo Dresses 1,Women,36.27,30.0,High Risk


# Returns Analysis

In [345]:
returns = pd.read_csv("data/returns.csv")

returns.head()

,Return_ID,Transaction_ID,Return_Date,Return_Qty,Return_Amount,Return_Reason
0,RT000001,TX0011901,2025-06-17,1,879.20,Size Issue
1,RT000002,TX0007960,2024-08-19,1,1359.15,Wrong Item
2,RT000003,TX0002403,2024-07-31,1,1439.10,Defective
3,RT000004,TX0006299,2024-06-08,1,594.15,Size Issue
4,RT000005,TX0011748,2024-06-04,1,479.20,Changed Mind


In [346]:
returns.shape

(900, 6)

In [347]:
returns.dtypes

Return_ID             str
Transaction_ID        str
Return_Date           str
Return_Qty          int64
Return_Amount     float64
Return_Reason         str
dtype: object

In [348]:
returns.isnull().sum()

Return_ID         0
Transaction_ID    0
Return_Date       0
Return_Qty        0
Return_Amount     0
Return_Reason     0
dtype: int64

In [349]:
returns.duplicated().sum()

np.int64(0)

In [350]:
# Return_Reason

returns["Return_Reason"].value_counts(dropna=False)

Return_Reason
Size Issue       287
Changed Mind     259
Quality Issue    153
Defective        125
Wrong Item        76
Name: count, dtype: int64

In [351]:
# convert the date into datetime

returns["Return_Date"] = pd.to_datetime(
    returns["Return_Date"]
)

In [352]:
# Validation

returns["Transaction_ID"].nunique()

900

In [353]:
returns["Return_ID"].nunique()

900

In [354]:
returns.head()
returns.shape
returns.dtypes
returns.isnull().sum()
returns.duplicated().sum()
returns["Return_Reason"].value_counts(dropna=False)

Return_Reason
Size Issue       287
Changed Mind     259
Quality Issue    153
Defective        125
Wrong Item        76
Name: count, dtype: int64

In [355]:
print("Shape:")
print(returns.shape)

print("\nData Types:")
print(returns.dtypes)

print("\nMissing Values:")
print(returns.isnull().sum())

print("\nDuplicate Rows:")
print(returns.duplicated().sum())

print("\nUnique Return IDs:")
print(returns["Return_ID"].nunique())

print("\nUnique Transaction IDs:")
print(returns["Transaction_ID"].nunique())

Shape:
(900, 6)

Data Types:
Return_ID                    str
Transaction_ID               str
Return_Date       datetime64[us]
Return_Qty                 int64
Return_Amount            float64
Return_Reason                str
dtype: object

Missing Values:
Return_ID         0
Transaction_ID    0
Return_Date       0
Return_Qty        0
Return_Amount     0
Return_Reason     0
dtype: int64

Duplicate Rows:
0

Unique Return IDs:
900

Unique Transaction IDs:
900


In [356]:
# Validate, all the returned transactions actually exist in the sales data

unmatched_returns = returns[
    ~returns["Transaction_ID"].isin(sales_analysis["Transaction_ID"])
]

unmatched_returns.shape

(0, 6)

## 1. Total Returned Units

In [357]:
total_returned_units = returns["Return_Qty"].sum()

print(f"Total Returned Units: {total_returned_units:,}")

Total Returned Units: 900


## 2. Unit Return Rate %

### Return Rate % = Returned Units / Units Sold × 100

In [358]:
return_rate_pct = (
    total_returned_units / total_units * 100
)

print(f"Return Rate: {return_rate_pct:.2f}%")

Return Rate: 3.19%


## 3. Return Transaction Rate %

In [359]:
return_transaction_rate = (
    returns["Transaction_ID"].nunique()
    / total_transactions
    * 100
)

print(
    f"Return Transaction Rate: "
    f"{return_transaction_rate:.2f}%"
)

Return Transaction Rate: 6.00%


## 4. Total Return Amount

In [360]:
total_return_amount = returns["Return_Amount"].sum()

print(
    f"Total Return Amount: "
    f"₹{total_return_amount:,.2f}"
)

Total Return Amount: ₹695,412.60


## 5. Return-adjusted Net Sales

In [361]:
adjusted_net_sales = (
    total_sales - total_return_amount
)

print(
    f"Return-Adjusted Net Sales: "
    f"₹{adjusted_net_sales:,.2f}"
)

Return-Adjusted Net Sales: ₹21,442,508.50


## Return Value Rate %

In [362]:
return_value_rate = (
    total_return_amount / total_sales * 100
)

print(f"Return Value Rate: {return_value_rate:.2f}%")

Return Value Rate: 3.14%


### Category-wise Return Analysis

In [363]:

returns_detail = returns.merge(
    sales_analysis[
        [
            "Transaction_ID",
            "Product_ID",
            "Product_Name",
            "Category",
            "Subcategory",
            "Store_ID",
            "Store_Name",
            "Region",
            "Quantity",
            "Net_Sales"
        ]
    ],
    on="Transaction_ID",
    how="left"
)

In [364]:
# Merge Validate

returns.shape


(900, 6)

In [365]:
returns_detail.shape

(900, 15)

In [366]:
returns_detail["Category"].isnull().sum()

np.int64(0)

## Category-wise returned units and amount

In [367]:
category_returns = (
    returns_detail
    .groupby("Category", as_index=False)
    .agg(
        Returned_Units=("Return_Qty", "sum"),
        Return_Transactions=("Transaction_ID", "nunique"),
        Return_Amount=("Return_Amount", "sum")
    )
)

category_returns

,Category,Returned_Units,Return_Transactions,Return_Amount
0,Accessories,183,183,144799.50
1,Footwear,166,166,136648.65
2,Kids,182,182,139038.15
3,Men,172,172,134102.40
4,Women,197,197,140823.90


In [368]:
# Merge with existing category_performance

category_return_analysis = category_performance.merge(
    category_returns,
    on="Category",
    how="left"
)

In [369]:
# proper return rates

category_return_analysis["Unit_Return_Rate_Pct"] = (
    category_return_analysis["Returned_Units"]
    / category_return_analysis["Units_Sold"]
    * 100
).round(2)

category_return_analysis["Transaction_Return_Rate_Pct"] = (
    category_return_analysis["Return_Transactions"]
    / category_return_analysis["Transactions"]
    * 100
).round(2)

category_return_analysis["Return_Value_Rate_Pct"] = (
    category_return_analysis["Return_Amount"]
    / category_return_analysis["Net_Sales"]
    * 100
).round(2)

In [370]:
category_return_analysis[[
    "Category",
    "Net_Sales",
    "Units_Sold",
    "Returned_Units",
    "Unit_Return_Rate_Pct",
    "Return_Transactions",
    "Transaction_Return_Rate_Pct",
    "Return_Amount",
    "Return_Value_Rate_Pct"
]].sort_values(
    "Unit_Return_Rate_Pct",
    ascending=False
)

,Category,Net_Sales,Units_Sold,Returned_Units,Unit_Return_Rate_Pct,Return_Transactions,Transaction_Return_Rate_Pct,Return_Amount,Return_Value_Rate_Pct
4,Women,4118079.20,5655,197,3.48,197,6.60,140823.90,3.42
0,Accessories,4713462.60,5696,183,3.21,183,6.03,144799.50,3.07
2,Kids,4354673.00,5693,182,3.20,182,6.12,139038.15,3.19
3,Men,4602179.85,5675,172,3.03,172,5.72,134102.40,2.91
1,Footwear,4349526.45,5525,166,3.00,166,5.54,136648.65,3.14


## Next — Return Reason by Category

In [371]:
return_reason_category = (
    returns_detail
    .groupby(
        ["Category", "Return_Reason"],
        as_index=False
    )
    .agg(
        Return_Count=("Return_ID", "nunique"),
        Return_Amount=("Return_Amount", "sum")
    )
)

return_reason_category

,Category,Return_Reason,Return_Count,Return_Amount
0,Accessories,Changed Mind,53,42789.70
1,Accessories,Defective,23,16565.95
2,Accessories,Quality Issue,27,22467.20
3,Accessories,Size Issue,64,50820.05
4,Accessories,Wrong Item,16,12156.60
5,Footwear,Changed Mind,47,40084.95
6,Footwear,Defective,28,24141.70
7,Footwear,Quality Issue,27,20326.90
8,Footwear,Size Issue,52,42884.90
9,Footwear,Wrong Item,12,9210.20


In [372]:
# Pivot Table

return_reason_pivot = pd.pivot_table(
    returns_detail,
    index="Category",
    columns="Return_Reason",
    values="Return_ID",
    aggfunc="nunique",
    fill_value=0
)

return_reason_pivot

Return_Reason,Changed Mind,Defective,Quality Issue,Size Issue,Wrong Item
Category,,,,,
Accessories,53,23,27,64,16
Footwear,47,28,27,52,12
Kids,57,25,31,52,17
Men,51,18,29,63,11
Women,51,31,39,56,20


In [373]:
return_reason_pct = (
    return_reason_pivot
    .div(return_reason_pivot.sum(axis=1), axis=0)
    * 100
).round(2)

return_reason_pct

Return_Reason,Changed Mind,Defective,Quality Issue,Size Issue,Wrong Item
Category,,,,,
Accessories,28.96,12.57,14.75,34.97,8.74
Footwear,28.31,16.87,16.27,31.33,7.23
Kids,31.32,13.74,17.03,28.57,9.34
Men,29.65,10.47,16.86,36.63,6.40
Women,25.89,15.74,19.80,28.43,10.15


### Product-level return analysis

In [374]:
product_returns = (
    returns_detail
    .groupby(
        ["Product_ID", "Product_Name", "Category", "Subcategory"],
        as_index=False
    )
    .agg(
        Returned_Units=("Return_Qty", "sum"),
        Return_Transactions=("Transaction_ID", "nunique"),
        Return_Amount=("Return_Amount", "sum")
    )
)

In [375]:
# Merge with product_performance

product_return_analysis = product_performance.merge(
    product_returns,
    on=[
        "Product_ID",
        "Product_Name",
        "Category",
        "Subcategory"
    ],
    how="left"
)

In [376]:
# For products where no return took place, the NaN is zero:

product_return_analysis[
    ["Returned_Units", "Return_Transactions", "Return_Amount"]
] = product_return_analysis[
    ["Returned_Units", "Return_Transactions", "Return_Amount"]
].fillna(0)

In [377]:
# product-level unit return rate

product_return_analysis["Unit_Return_Rate_Pct"] = (
    product_return_analysis["Returned_Units"]
    / product_return_analysis["Units_Sold"]
    * 100
).round(2)

In [378]:
# return value rate

product_return_analysis["Return_Value_Rate_Pct"] = (
    product_return_analysis["Return_Amount"]
    / product_return_analysis["Net_Sales"]
    * 100
).round(2)

In [379]:
# Finally highest-return products:

product_return_analysis[[
    "Product_ID",
    "Product_Name",
    "Category",
    "Subcategory",
    "Units_Sold",
    "Returned_Units",
    "Unit_Return_Rate_Pct",
    "Return_Amount",
    "Return_Value_Rate_Pct"
]].sort_values(
    "Unit_Return_Rate_Pct",
    ascending=False
).head(15)

,Product_ID,Product_Name,Category,Subcategory,Units_Sold,Returned_Units,Unit_Return_Rate_Pct,Return_Amount,Return_Value_Rate_Pct
55,PR0056,NovaWear Belts 4,Accessories,Belts,443,23,5.19,5636.15,5.02
21,PR0022,Everyday Dresses 2,Women,Dresses,484,24,4.96,21870.10,4.84
58,PR0059,PrimeFit Wallets 3,Accessories,Wallets,514,24,4.67,12399.30,4.71
18,PR0019,StyleHub Tops 3,Women,Tops,500,23,4.60,15380.75,4.50
0,PR0001,StyleHub Shirts 1,Men,Shirts,438,20,4.57,6543.60,4.41
19,PR0020,NovaWear Tops 4,Women,Tops,428,18,4.21,13395.10,4.13
42,PR0043,StyleHub Sports 3,Footwear,Sports,500,21,4.20,6982.50,4.13
52,PR0053,PrimeFit Belts 1,Accessories,Belts,462,19,4.11,23083.50,4.21
16,PR0017,PrimeFit Tops 1,Women,Tops,506,20,3.95,10153.05,3.93
7,PR0008,NovaWear T-Shirts 4,Men,T-Shirts,486,19,3.91,17529.05,3.87


## Product-wise dominant return reason

In [380]:
product_return_reason = (
    returns_detail
    .groupby(
        ["Product_ID", "Product_Name", "Return_Reason"],
        as_index=False
    )
    .agg(
        Return_Count=("Return_ID", "nunique")
    )
)

In [381]:
# top return reason

top_return_reason = (
    product_return_reason
    .sort_values(
        ["Product_ID", "Return_Count"],
        ascending=[True, False]
    )
    .drop_duplicates("Product_ID")
)

top_return_reason.head(15)

,Product_ID,Product_Name,Return_Reason,Return_Count
3,PR0001,StyleHub Shirts 1,Size Issue,8
5,PR0002,NovaWear Shirts 2,Changed Mind,5
11,PR0003,TrendCo Shirts 3,Size Issue,4
15,PR0004,Everyday Shirts 4,Size Issue,6
19,PR0005,PrimeFit T-Shirts 1,Size Issue,5
23,PR0006,UrbanEdge T-Shirts 2,Quality Issue,4
28,PR0007,StyleHub T-Shirts 3,Size Issue,8
30,PR0008,NovaWear T-Shirts 4,Changed Mind,6
38,PR0009,TrendCo Jeans 1,Size Issue,7
42,PR0010,Everyday Jeans 2,Size Issue,5


In [382]:
# merge with product return analysis

product_return_hotspots = product_return_analysis.merge(
    top_return_reason[
        ["Product_ID", "Return_Reason", "Return_Count"]
    ],
    on="Product_ID",
    how="left"
)

In [383]:
# top 15

product_return_hotspots[[
    "Product_ID",
    "Product_Name",
    "Category",
    "Unit_Return_Rate_Pct",
    "Return_Amount",
    "Return_Value_Rate_Pct",
    "Return_Reason",
    "Return_Count"
]].sort_values(
    "Unit_Return_Rate_Pct",
    ascending=False
).head(15)

,Product_ID,Product_Name,Category,Unit_Return_Rate_Pct,Return_Amount,Return_Value_Rate_Pct,Return_Reason,Return_Count
55,PR0056,NovaWear Belts 4,Accessories,5.19,5636.15,5.02,Changed Mind,6
21,PR0022,Everyday Dresses 2,Women,4.96,21870.10,4.84,Changed Mind,9
58,PR0059,PrimeFit Wallets 3,Accessories,4.67,12399.30,4.71,Size Issue,10
18,PR0019,StyleHub Tops 3,Women,4.60,15380.75,4.50,Changed Mind,6
0,PR0001,StyleHub Shirts 1,Men,4.57,6543.60,4.41,Size Issue,8
19,PR0020,NovaWear Tops 4,Women,4.21,13395.10,4.13,Changed Mind,4
42,PR0043,StyleHub Sports 3,Footwear,4.20,6982.50,4.13,Size Issue,7
52,PR0053,PrimeFit Belts 1,Accessories,4.11,23083.50,4.21,Size Issue,9
16,PR0017,PrimeFit Tops 1,Women,3.95,10153.05,3.93,Defective,7
7,PR0008,NovaWear T-Shirts 4,Men,3.91,17529.05,3.87,Changed Mind,6


In [384]:
inventory = pd.read_csv("data/inventory_monthly.csv")

inventory.head()

,Snapshot_Month,Store_ID,Product_ID,Opening_Stock,Receipts,Units_Sold,Closing_Stock,Stockout_Days,Inventory_Value
0,2025-01-01,ST001,PR0001,19,36,23,32,0,7669.44
1,2025-01-01,ST001,PR0002,9,41,23,27,0,6271.83
2,2025-01-01,ST001,PR0003,32,50,7,75,0,23383.50
3,2025-01-01,ST001,PR0004,58,55,7,106,0,37722.22
4,2025-01-01,ST001,PR0005,6,45,25,26,0,12922.00


In [385]:
print("Shape:")
print(inventory.shape)

Shape:
(17280, 9)


In [386]:
print("\nData Types:")
print(inventory.dtypes)


Data Types:
Snapshot_Month         str
Store_ID               str
Product_ID             str
Opening_Stock        int64
Receipts             int64
Units_Sold           int64
Closing_Stock        int64
Stockout_Days        int64
Inventory_Value    float64
dtype: object


In [387]:
print("\nMissing Values:")
print(inventory.isnull().sum())


Missing Values:
Snapshot_Month     0
Store_ID           0
Product_ID         0
Opening_Stock      0
Receipts           0
Units_Sold         0
Closing_Stock      0
Stockout_Days      0
Inventory_Value    0
dtype: int64


In [388]:
print("\nDuplicates:")
print(inventory.duplicated().sum())


Duplicates:
0


In [389]:
# inventory month

inventory["Snapshot_Month"] = pd.to_datetime(
    inventory["Snapshot_Month"]
)

In [390]:
print("Min Month:", inventory["Snapshot_Month"].min())
print("Max Month:", inventory["Snapshot_Month"].max())

Min Month: 2025-01-01 00:00:00
Max Month: 2025-12-01 00:00:00


In [391]:
# basic business validation

inventory[
    inventory["Closing_Stock"] !=
    (
        inventory["Opening_Stock"]
        + inventory["Receipts"]
        - inventory["Units_Sold"]
    )
].shape

(0, 9)

## Inventory KPI 1 — Sell-Through %

In [392]:
inventory["Available_Units"] = (
    inventory["Opening_Stock"]
    + inventory["Receipts"]
)

In [393]:
inventory["Sell_Through_Pct"] = (
    inventory["Units_Sold"]
    / inventory["Available_Units"]
    * 100
).round(2)

In [394]:
inventory[[
    "Snapshot_Month",
    "Store_ID",
    "Product_ID",
    "Opening_Stock",
    "Receipts",
    "Available_Units",
    "Units_Sold",
    "Closing_Stock",
    "Sell_Through_Pct"
]].head(10)

,Snapshot_Month,Store_ID,Product_ID,Opening_Stock,Receipts,Available_Units,Units_Sold,Closing_Stock,Sell_Through_Pct
0,2025-01-01,ST001,PR0001,19,36,55,23,32,41.82
1,2025-01-01,ST001,PR0002,9,41,50,23,27,46.00
2,2025-01-01,ST001,PR0003,32,50,82,7,75,8.54
3,2025-01-01,ST001,PR0004,58,55,113,7,106,6.19
4,2025-01-01,ST001,PR0005,6,45,51,25,26,49.02
5,2025-01-01,ST001,PR0006,49,17,66,33,33,50.00
6,2025-01-01,ST001,PR0007,58,32,90,0,90,0.00
7,2025-01-01,ST001,PR0008,48,48,96,16,80,16.67
8,2025-01-01,ST001,PR0009,57,42,99,48,51,48.48
9,2025-01-01,ST001,PR0010,33,51,84,4,80,4.76


### Company-level Sell-Through

In [395]:
overall_sell_through = (
    inventory["Units_Sold"].sum()
    / inventory["Available_Units"].sum()
    * 100
)

print(f"Overall Sell-Through %: {overall_sell_through:.2f}%")

Overall Sell-Through %: 33.77%


## Inventory KPI 2 — Stock-out Rate

In [396]:
inventory["Days_In_Month"] = (
    inventory["Snapshot_Month"].dt.days_in_month
)

In [397]:
inventory["Stockout_Rate_Pct"] = (
    inventory["Stockout_Days"]
    / inventory["Days_In_Month"]
    * 100
).round(2)

In [398]:
overall_stockout_rate = (
    inventory["Stockout_Days"].sum()
    / inventory["Days_In_Month"].sum()
    * 100
)

print(f"Overall Stock-out Rate: {overall_stockout_rate:.2f}%")

Overall Stock-out Rate: 1.95%


In [399]:
print(f"Overall Sell-Through %: {overall_sell_through:.2f}%")
print(f"Overall Stock-out Rate: {overall_stockout_rate:.2f}%")

Overall Sell-Through %: 33.77%
Overall Stock-out Rate: 1.95%


#### Step 1 — Latest month

In [406]:
latest_month = inventory["Snapshot_Month"].max()

latest_month

Timestamp('2025-12-01 00:00:00')

#### Step 2 — Latest inventory snapshot

In [407]:
latest_inventory = inventory[
    inventory["Snapshot_Month"] == latest_month
].copy()

latest_inventory.shape

(1440, 13)

#### Step 3 — Current Inventory Value

In [408]:
current_inventory_value = (
    latest_inventory["Inventory_Value"].sum()
)

print(
    f"Current Inventory Value: "
    f"₹{current_inventory_value:,.2f}"
)

Current Inventory Value: ₹29,711,691.83


#### Step 4 — Slow-Moving Inventory definition

In [409]:
latest_inventory["Slow_Moving_Flag"] = np.where(
    (
        (latest_inventory["Sell_Through_Pct"] < 20)
        &
        (latest_inventory["Closing_Stock"] > 0)
    ),
    "Slow Moving",
    "Normal"
)

In [410]:
latest_inventory[
    "Slow_Moving_Flag"
].value_counts()

Slow_Moving_Flag
Normal         1117
Slow Moving     323
Name: count, dtype: int64

#### Step 5 — Slow-moving inventory value

In [411]:
slow_moving_value = (
    latest_inventory.loc[
        latest_inventory["Slow_Moving_Flag"] == "Slow Moving",
        "Inventory_Value"
    ].sum()
)

print(
    f"Slow-Moving Inventory Value: "
    f"₹{slow_moving_value:,.2f}"
)

Slow-Moving Inventory Value: ₹10,759,584.55


#### Step 6 — Slow-Moving Inventory %

In [412]:
slow_moving_inventory_pct = (
    slow_moving_value
    / current_inventory_value
    * 100
)

print(
    f"Slow-Moving Inventory %: "
    f"{slow_moving_inventory_pct:.2f}%"
)

Slow-Moving Inventory %: 36.21%


## Next — Category-wise slow-moving inventory

In [413]:
latest_inventory_analysis = latest_inventory.merge(
    products[
        [
            "Product_ID",
            "Product_Name",
            "Category",
            "Subcategory",
            "Brand"
        ]
    ],
    on="Product_ID",
    how="left"
)

In [414]:
latest_inventory_analysis.shape

(1440, 18)

In [415]:
latest_inventory_analysis[
    latest_inventory_analysis["Category"].isnull()
]["Product_ID"].unique()

<ArrowStringArray>
[]
Length: 0, dtype: str

In [416]:
# category summary:

category_inventory = (
    latest_inventory_analysis
    .groupby("Category", as_index=False)
    .agg(
        Inventory_Value=("Inventory_Value", "sum"),
        Closing_Stock=("Closing_Stock", "sum"),
        Units_Sold=("Units_Sold", "sum"),
        Available_Units=("Available_Units", "sum")
    )
)

In [417]:
# Category Sell-Through:

category_inventory["Sell_Through_Pct"] = (
    category_inventory["Units_Sold"]
    / category_inventory["Available_Units"]
    * 100
).round(2)

In [418]:
# slow-moving value separately

category_slow_moving = (
    latest_inventory_analysis[
        latest_inventory_analysis["Slow_Moving_Flag"] == "Slow Moving"
    ]
    .groupby("Category", as_index=False)
    .agg(
        Slow_Moving_Inventory_Value=("Inventory_Value", "sum"),
        Slow_Moving_SKUs=("Product_ID", "count")
    )
)

In [419]:
category_inventory_analysis = category_inventory.merge(
    category_slow_moving,
    on="Category",
    how="left"
)

In [420]:
# Slow-moving value percentage:

category_inventory_analysis["Slow_Moving_Value_Pct"] = (
    category_inventory_analysis["Slow_Moving_Inventory_Value"]
    / category_inventory_analysis["Inventory_Value"]
    * 100
).round(2)

In [421]:
category_inventory_analysis.sort_values(
    "Slow_Moving_Value_Pct",
    ascending=False
)

,Category,Inventory_Value,Closing_Stock,Units_Sold,Available_Units,Sell_Through_Pct,Slow_Moving_Inventory_Value,Slow_Moving_SKUs,Slow_Moving_Value_Pct
4,Women,5638013.00,11809,5849,17658,33.12,2285521.93,72,40.54
1,Footwear,6035256.84,11686,6171,17857,34.56,2341356.97,66,38.79
0,Accessories,6148660.84,11712,6270,17982,34.87,2261126.15,65,36.77
3,Men,5703065.78,11402,6119,17521,34.92,1906574.36,63,33.43
2,Kids,6186695.37,11864,6163,18027,34.19,1965005.14,57,31.76


## opposite problem: Stock-out / Replenishment Opportunity

In [422]:
product_inventory = (
    latest_inventory_analysis
    .groupby(
        ["Product_ID", "Product_Name", "Category", "Subcategory"],
        as_index=False
    )
    .agg(
        Available_Units=("Available_Units", "sum"),
        Units_Sold=("Units_Sold", "sum"),
        Closing_Stock=("Closing_Stock", "sum"),
        Stockout_Days=("Stockout_Days", "sum"),
        Inventory_Value=("Inventory_Value", "sum"),
        Days_In_Month=("Days_In_Month", "sum")
    )
)

In [423]:
# product-level Sell-Through:

product_inventory["Sell_Through_Pct"] = (
    product_inventory["Units_Sold"]
    / product_inventory["Available_Units"]
    * 100
).round(2)

In [424]:
product_inventory["Stockout_Rate_Pct"] = (
    product_inventory["Stockout_Days"]
    / product_inventory["Days_In_Month"]
    * 100
).round(2)

In [425]:
replenishment_hotspots = product_inventory[
    (product_inventory["Sell_Through_Pct"] >= 40)
    &
    (product_inventory["Stockout_Days"] > 0)
].copy()

In [426]:
replenishment_hotspots = (
    replenishment_hotspots
    .sort_values(
        ["Stockout_Rate_Pct", "Sell_Through_Pct"],
        ascending=False
    )
)

replenishment_hotspots.head(15)

,Product_ID,Product_Name,Category,Subcategory,Available_Units,Units_Sold,Closing_Stock,Stockout_Days,Inventory_Value,Days_In_Month,Sell_Through_Pct,Stockout_Rate_Pct
48,PR0049,StyleHub Bags 1,Accessories,Bags,1394,637,757,41,453632.25,744,45.70,5.51
11,PR0012,UrbanEdge Jeans 4,Men,Jeans,1388,569,819,30,701227.80,744,40.99,4.03
32,PR0033,TrendCo Sets 1,Kids,Sets,1308,533,775,28,357895.00,744,40.75,3.76
53,PR0054,UrbanEdge Belts 2,Accessories,Belts,1338,553,785,5,783767.55,744,41.33,0.67


## Store-Level Replenishment Drill-down

In [427]:
hotspot_product_ids = replenishment_hotspots["Product_ID"].tolist()

store_replenishment = (
    latest_inventory_analysis[
        latest_inventory_analysis["Product_ID"].isin(hotspot_product_ids)
    ]
    .sort_values(
        ["Product_ID", "Stockout_Days"],
        ascending=[True, False]
    )
)

store_replenishment[[
    "Store_ID",
    "Product_ID",
    "Product_Name",
    "Category",
    "Available_Units",
    "Units_Sold",
    "Closing_Stock",
    "Stockout_Days",
    "Sell_Through_Pct"
]].head(30)

,Store_ID,Product_ID,Product_Name,Category,Available_Units,Units_Sold,Closing_Stock,Stockout_Days,Sell_Through_Pct
131,ST003,PR0012,UrbanEdge Jeans 4,Men,24,24,0,11,100.00
971,ST017,PR0012,UrbanEdge Jeans 4,Men,7,7,0,10,100.00
911,ST016,PR0012,UrbanEdge Jeans 4,Men,13,13,0,9,100.00
11,ST001,PR0012,UrbanEdge Jeans 4,Men,37,14,23,0,37.84
71,ST002,PR0012,UrbanEdge Jeans 4,Men,66,33,33,0,50.00
191,ST004,PR0012,UrbanEdge Jeans 4,Men,81,20,61,0,24.69
251,ST005,PR0012,UrbanEdge Jeans 4,Men,58,23,35,0,39.66
311,ST006,PR0012,UrbanEdge Jeans 4,Men,52,42,10,0,80.77
371,ST007,PR0012,UrbanEdge Jeans 4,Men,88,16,72,0,18.18
431,ST008,PR0012,UrbanEdge Jeans 4,Men,92,22,70,0,23.91


## Next major KPI — Inventory Turnover

#### Step 1 — 2025 COGS

In [428]:
sales_2025 = sales_analysis[
    sales_analysis["Year"] == 2025
]

cogs_2025 = sales_2025["Cost_Amount"].sum()

print(f"2025 COGS: ₹{cogs_2025:,.2f}")

2025 COGS: ₹7,420,260.56


#### Step 2 — Monthly total inventory value

In [429]:
monthly_inventory_value = (
    inventory
    .groupby("Snapshot_Month", as_index=False)
    .agg(
        Inventory_Value=("Inventory_Value", "sum")
    )
)

monthly_inventory_value

,Snapshot_Month,Inventory_Value
0,2025-01-01,30962669.63
1,2025-02-01,30085501.20
2,2025-03-01,30341977.48
3,2025-04-01,30482043.74
4,2025-05-01,31315950.89
5,2025-06-01,30706101.49
6,2025-07-01,31083291.53
7,2025-08-01,30853410.26
8,2025-09-01,30784071.25
9,2025-10-01,30916449.84


#### Step 3 — Average Inventory Value

In [430]:
average_inventory_value = (
    monthly_inventory_value["Inventory_Value"].mean()
)

print(
    f"Average Inventory Value: "
    f"₹{average_inventory_value:,.2f}"
)

Average Inventory Value: ₹30,726,557.77


#### Step 4 — Inventory Turnover

In [431]:
# Inventory Turnover = COGS ÷ Average Inventory Value

inventory_turnover = (
    cogs_2025 / average_inventory_value
)

print(
    f"Inventory Turnover: "
    f"{inventory_turnover:.2f} times"
)

Inventory Turnover: 0.24 times


In [432]:
inventory_turnover_data = inventory.merge(
    products[
        ["Product_ID", "Unit_Cost"]
    ],
    on="Product_ID",
    how="left"
)

In [433]:
inventory_turnover_data["Unit_Cost"].isnull().sum()

np.int64(0)

In [434]:
inventory_turnover_data["Inventory_COGS"] = (
    inventory_turnover_data["Units_Sold"]
    * inventory_turnover_data["Unit_Cost"]
)

In [435]:
inventory_cogs_2025 = (
    inventory_turnover_data["Inventory_COGS"].sum()
)

print(
    f"Inventory-based 2025 COGS: "
    f"₹{inventory_cogs_2025:,.2f}"
)

Inventory-based 2025 COGS: ₹188,563,255.62


In [436]:
inventory_turnover = (
    inventory_cogs_2025
    / average_inventory_value
)

print(
    f"Inventory Turnover: "
    f"{inventory_turnover:.2f} times"
)

Inventory Turnover: 6.14 times


#### Days Inventory Outstanding

In [437]:
days_inventory = (
    365 / inventory_turnover
)

print(
    f"Days Inventory Outstanding: "
    f"{days_inventory:.2f} days"
)

Days Inventory Outstanding: 59.48 days


# Target vs Actual

### Step 1 — 2024 store-month actual sales

In [438]:
sales_2024 = (
    sales_analysis[sales_analysis["Year"] == 2024]
    .groupby(
        ["Store_ID", "Month"],
        as_index=False
    )
    .agg(
        Previous_Year_Sales=("Net_Sales", "sum")
    )
)

sales_2024.head()

,Store_ID,Month,Previous_Year_Sales
0,ST001,1,36282.60
1,ST001,2,27175.45
2,ST001,3,41745.40
3,ST001,4,49553.55
4,ST001,5,36043.90


### Step 2 — Planned growth target assign

In [439]:
conditions = [
    sales_2024["Month"].isin([10, 11, 12]),
    sales_2024["Month"].isin([3, 8])
]

choices = [
    10,   # festive / peak months
    8
]

sales_2024["Target_Growth_Pct"] = np.select(
    conditions,
    choices,
    default=5
)

### Step 3 — 2025 Sales Target

In [440]:
sales_2024["Sales_Target"] = (
    sales_2024["Previous_Year_Sales"]
    * (1 + sales_2024["Target_Growth_Pct"] / 100)
).round(2)

### Step 4 — 2025 actual sales

In [441]:
sales_2025_monthly = (
    sales_analysis[sales_analysis["Year"] == 2025]
    .groupby(
        ["Store_ID", "Month"],
        as_index=False
    )
    .agg(
        Actual_Sales=("Net_Sales", "sum"),
        Actual_Profit=("Gross_Profit", "sum")
    )
)

### Step 5 — Target vs Actual merge

In [442]:
target_vs_actual = sales_2025_monthly.merge(
    sales_2024[
        [
            "Store_ID",
            "Month",
            "Previous_Year_Sales",
            "Target_Growth_Pct",
            "Sales_Target"
        ]
    ],
    on=["Store_ID", "Month"],
    how="left"
)

In [443]:
target_vs_actual.shape

(288, 7)

In [444]:
target_vs_actual.isnull().sum()

Store_ID               0
Month                  0
Actual_Sales           0
Actual_Profit          0
Previous_Year_Sales    0
Target_Growth_Pct      0
Sales_Target           0
dtype: int64

### Step 6 — Sales Variance ₹

In [445]:
target_vs_actual["Sales_Variance"] = (
    target_vs_actual["Actual_Sales"]
    - target_vs_actual["Sales_Target"]
).round(2)

### Step 7 — Target Achievement %

In [446]:
target_vs_actual["Target_Achievement_Pct"] = (
    target_vs_actual["Actual_Sales"]
    / target_vs_actual["Sales_Target"]
    * 100
).round(2)

### Step 8 — Variance %

In [447]:
target_vs_actual["Sales_Variance_Pct"] = (
    target_vs_actual["Sales_Variance"]
    / target_vs_actual["Sales_Target"]
    * 100
).round(2)

### Step 9 — Target Status

In [448]:
target_vs_actual["Target_Status"] = np.where(
    target_vs_actual["Actual_Sales"]
    >= target_vs_actual["Sales_Target"],
    "Achieved",
    "Missed"
)

In [449]:
target_vs_actual.head(10)

,Store_ID,Month,Actual_Sales,Actual_Profit,Previous_Year_Sales,Target_Growth_Pct,Sales_Target,Sales_Variance,Target_Achievement_Pct,Sales_Variance_Pct,Target_Status
0,ST001,1,45161.10,15697.98,36282.60,5,38096.73,7064.37,118.54,18.54,Achieved
1,ST001,2,43022.10,16710.75,27175.45,5,28534.22,14487.88,150.77,50.77,Achieved
2,ST001,3,29947.05,9987.73,41745.40,8,45085.03,-15137.98,66.42,-33.58,Missed
3,ST001,4,35479.95,9012.27,49553.55,5,52031.23,-16551.28,68.19,-31.81,Missed
4,ST001,5,38019.60,11997.12,36043.90,5,37846.10,173.50,100.46,0.46,Achieved
5,ST001,6,45185.85,14104.10,48395.80,5,50815.59,-5629.74,88.92,-11.08,Missed
6,ST001,7,41544.40,13600.28,37867.30,5,39760.67,1783.73,104.49,4.49,Achieved
7,ST001,8,42477.95,14640.50,28938.55,8,31253.63,11224.32,135.91,35.91,Achieved
8,ST001,9,48597.40,17995.56,50120.05,5,52626.05,-4028.65,92.34,-7.66,Missed
9,ST001,10,20366.40,7229.68,46713.95,10,51385.34,-31018.94,39.63,-60.37,Missed


### Step 10 — How many store-months targets were achieved?

In [450]:
target_vs_actual["Target_Status"].value_counts()

Target_Status
Missed      168
Achieved    120
Name: count, dtype: int64

In [451]:
achievement_rate = (
    (target_vs_actual["Target_Status"] == "Achieved").mean()
    * 100
)

print(f"Target Achievement Rate: {achievement_rate:.2f}%")

Target Achievement Rate: 41.67%


In [452]:
target_vs_actual.head(10)

target_vs_actual["Target_Status"].value_counts()

print(f"Target Achievement Rate: {achievement_rate:.2f}%")

Target Achievement Rate: 41.67%


In [453]:
target_vs_actual.head(10)

,Store_ID,Month,Actual_Sales,Actual_Profit,Previous_Year_Sales,Target_Growth_Pct,Sales_Target,Sales_Variance,Target_Achievement_Pct,Sales_Variance_Pct,Target_Status
0,ST001,1,45161.10,15697.98,36282.60,5,38096.73,7064.37,118.54,18.54,Achieved
1,ST001,2,43022.10,16710.75,27175.45,5,28534.22,14487.88,150.77,50.77,Achieved
2,ST001,3,29947.05,9987.73,41745.40,8,45085.03,-15137.98,66.42,-33.58,Missed
3,ST001,4,35479.95,9012.27,49553.55,5,52031.23,-16551.28,68.19,-31.81,Missed
4,ST001,5,38019.60,11997.12,36043.90,5,37846.10,173.50,100.46,0.46,Achieved
5,ST001,6,45185.85,14104.10,48395.80,5,50815.59,-5629.74,88.92,-11.08,Missed
6,ST001,7,41544.40,13600.28,37867.30,5,39760.67,1783.73,104.49,4.49,Achieved
7,ST001,8,42477.95,14640.50,28938.55,8,31253.63,11224.32,135.91,35.91,Achieved
8,ST001,9,48597.40,17995.56,50120.05,5,52626.05,-4028.65,92.34,-7.66,Missed
9,ST001,10,20366.40,7229.68,46713.95,10,51385.34,-31018.94,39.63,-60.37,Missed


# company-level Target vs Actual KPI

In [454]:
company_target = target_vs_actual["Sales_Target"].sum()
company_actual = target_vs_actual["Actual_Sales"].sum()

company_variance = company_actual - company_target

company_target_achievement = (
    company_actual / company_target * 100
)

print(f"Company Sales Target: ₹{company_target:,.2f}")
print(f"Company Actual Sales: ₹{company_actual:,.2f}")
print(f"Sales Variance: ₹{company_variance:,.2f}")
print(
    f"Company Target Achievement: "
    f"{company_target_achievement:.2f}%"
)

Company Sales Target: ₹11,673,413.61
Company Actual Sales: ₹11,196,493.50
Sales Variance: ₹-476,920.11
Company Target Achievement: 95.91%


## Next — Monthly Target Performance

In [455]:
monthly_target_performance = (
    target_vs_actual
    .groupby("Month", as_index=False)
    .agg(
        Sales_Target=("Sales_Target", "sum"),
        Actual_Sales=("Actual_Sales", "sum")
    )
)

monthly_target_performance["Variance"] = (
    monthly_target_performance["Actual_Sales"]
    - monthly_target_performance["Sales_Target"]
)

monthly_target_performance["Target_Achievement_Pct"] = (
    monthly_target_performance["Actual_Sales"]
    / monthly_target_performance["Sales_Target"]
    * 100
).round(2)

monthly_target_performance

,Month,Sales_Target,Actual_Sales,Variance,Target_Achievement_Pct
0,1,924044.87,1001269.65,77224.78,108.36
1,2,962482.71,883701.15,-78781.56,91.81
2,3,924616.30,1011771.80,87155.50,109.43
3,4,1005911.00,917881.95,-88029.05,91.25
4,5,1036818.54,924695.70,-112122.84,89.19
5,6,1000120.92,941871.75,-58249.17,94.18
6,7,974351.78,1013047.60,38695.82,103.97
7,8,1034299.84,1046459.15,12159.31,101.18
8,9,939764.69,856794.85,-82969.84,91.17
9,10,969999.77,872344.35,-97655.42,89.93


### Store-level Target Performance

In [456]:
store_target_performance = (
    target_vs_actual
    .groupby("Store_ID", as_index=False)
    .agg(
        Sales_Target=("Sales_Target", "sum"),
        Actual_Sales=("Actual_Sales", "sum"),
        Months_Achieved=(
            "Target_Status",
            lambda x: (x == "Achieved").sum()
        )
    )
)

In [457]:
store_target_performance["Sales_Variance"] = (
    store_target_performance["Actual_Sales"]
    - store_target_performance["Sales_Target"]
).round(2)

In [458]:
store_target_performance["Target_Achievement_Pct"] = (
    store_target_performance["Actual_Sales"]
    / store_target_performance["Sales_Target"]
    * 100
).round(2)

In [459]:
store_target_performance["Months_Missed"] = (
    12 - store_target_performance["Months_Achieved"]
)

In [460]:
store_target_performance = store_target_performance.merge(
    stores[
        ["Store_ID", "Store_Name", "City", "Region"]
    ],
    on="Store_ID",
    how="left"
)

In [461]:
# Top target performers

top_target_stores = (
    store_target_performance
    .sort_values(
        "Target_Achievement_Pct",
        ascending=False
    )
    .head(10)
)

top_target_stores

,Store_ID,Sales_Target,Actual_Sales,Months_Achieved,Sales_Variance,Target_Achievement_Pct,Months_Missed,Store_Name,City,Region
23,ST024,426862.87,529898.25,9,103035.38,124.14,3,Patna Store 2,Patna,East
2,ST003,430227.61,505498.60,8,75270.99,117.50,4,Lucknow Store 1,Lucknow,North
20,ST021,453848.73,499003.00,7,45154.27,109.95,5,Bhubaneswar Store 1,Bhubaneswar,East
5,ST006,444723.20,485085.75,6,40362.55,109.08,6,Jaipur Store 2,Jaipur,North
9,ST010,454718.47,488001.80,5,33283.33,107.32,7,Pune Store 2,Pune,West
18,ST019,448806.32,479142.70,7,30336.38,106.76,5,Kolkata Store 1,Kolkata,East
1,ST002,436718.49,446530.70,6,9812.21,102.25,6,Delhi Store 2,Delhi,North
3,ST004,516322.32,527230.90,4,10908.58,102.11,8,Lucknow Store 2,Lucknow,North
19,ST020,499574.94,490558.25,9,-9016.69,98.20,3,Kolkata Store 2,Kolkata,East
10,ST011,469536.87,454261.30,5,-15275.57,96.75,7,Ahmedabad Store 1,Ahmedabad,West


In [462]:
# Bottom target performers

bottom_target_stores = (
    store_target_performance
    .sort_values(
        "Target_Achievement_Pct"
    )
    .head(10)
)

bottom_target_stores

,Store_ID,Sales_Target,Actual_Sales,Months_Achieved,Sales_Variance,Target_Achievement_Pct,Months_Missed,Store_Name,City,Region
8,ST009,581543.59,445096.75,3,-136446.84,76.54,9,Pune Store 1,Pune,West
14,ST015,581887.81,456690.75,2,-125197.06,78.48,10,Hyderabad Store 1,Hyderabad,South
12,ST013,532923.94,441079.20,1,-91844.74,82.77,11,Bengaluru Store 1,Bengaluru,South
22,ST023,523488.86,440624.45,2,-82864.41,84.17,10,Patna Store 1,Patna,East
21,ST022,536240.70,466147.35,3,-70093.35,86.93,9,Bhubaneswar Store 2,Bhubaneswar,East
11,ST012,437526.85,382969.85,4,-54557.00,87.53,8,Ahmedabad Store 2,Ahmedabad,West
0,ST001,508741.21,454975.20,5,-53766.01,89.43,7,Delhi Store 1,Delhi,North
15,ST016,518507.04,475113.70,4,-43393.34,91.63,8,Hyderabad Store 2,Hyderabad,South
6,ST007,496910.73,461595.35,4,-35315.38,92.89,8,Mumbai Store 1,Mumbai,West
17,ST018,474490.75,445965.90,6,-28524.85,93.99,6,Chennai Store 2,Chennai,South


## cross-KPI comparison 2025-only

In [463]:
store_performance_2025 = (
    sales_2025
    .groupby(
        ["Store_ID", "Store_Name", "City", "Region"],
        as_index=False
    )
    .agg(
        Net_Sales=("Net_Sales", "sum"),
        Gross_Profit=("Gross_Profit", "sum"),
        Transactions=("Transaction_ID", "nunique"),
        Units_Sold=("Quantity", "sum")
    )
)

In [464]:
store_performance_2025["ATV"] = (
    store_performance_2025["Net_Sales"]
    / store_performance_2025["Transactions"]
).round(2)

store_performance_2025["UPT"] = (
    store_performance_2025["Units_Sold"]
    / store_performance_2025["Transactions"]
).round(2)

store_performance_2025["Gross_Margin_Pct"] = (
    store_performance_2025["Gross_Profit"]
    / store_performance_2025["Net_Sales"]
    * 100
).round(2)

In [465]:
cross_store_analysis = store_performance_2025.merge(
    store_target_performance[
        [
            "Store_ID",
            "Sales_Target",
            "Sales_Variance",
            "Target_Achievement_Pct",
            "Months_Achieved",
            "Months_Missed"
        ]
    ],
    on="Store_ID",
    how="left"
)

In [466]:
cross_store_analysis[[
    "Store_ID",
    "Store_Name",
    "Net_Sales",
    "Gross_Margin_Pct",
    "Transactions",
    "ATV",
    "UPT",
    "Sales_Target",
    "Target_Achievement_Pct",
    "Months_Achieved",
    "Months_Missed"
]].sort_values(
    "Target_Achievement_Pct",
    ascending=False
)

,Store_ID,Store_Name,Net_Sales,Gross_Margin_Pct,Transactions,ATV,UPT,Sales_Target,Target_Achievement_Pct,Months_Achieved,Months_Missed
23,ST024,Patna Store 2,529898.25,33.76,357,1484.31,1.83,426862.87,124.14,9,3
2,ST003,Lucknow Store 1,505498.60,33.11,337,1500.00,1.90,430227.61,117.50,8,4
20,ST021,Bhubaneswar Store 1,499003.00,34.14,322,1549.70,1.97,453848.73,109.95,7,5
5,ST006,Jaipur Store 2,485085.75,32.73,311,1559.76,1.96,444723.20,109.08,6,6
9,ST010,Pune Store 2,488001.80,35.12,337,1448.08,1.88,454718.47,107.32,5,7
18,ST019,Kolkata Store 1,479142.70,33.08,323,1483.41,1.84,448806.32,106.76,7,5
1,ST002,Delhi Store 2,446530.70,32.76,315,1417.56,1.80,436718.49,102.25,6,6
3,ST004,Lucknow Store 2,527230.90,35.23,341,1546.13,1.92,516322.32,102.11,4,8
19,ST020,Kolkata Store 2,490558.25,35.04,309,1587.57,1.97,499574.94,98.20,9,3
10,ST011,Ahmedabad Store 1,454261.30,35.01,302,1504.18,1.86,469536.87,96.75,5,7


## final store classification

In [467]:
company_margin_2025 = (
    sales_2025["Gross_Profit"].sum()
    / sales_2025["Net_Sales"].sum()
    * 100
)

company_atv_2025 = (
    sales_2025["Net_Sales"].sum()
    / sales_2025["Transaction_ID"].nunique()
)

company_upt_2025 = (
    sales_2025["Quantity"].sum()
    / sales_2025["Transaction_ID"].nunique()
)

print(f"2025 Company Margin: {company_margin_2025:.2f}%")
print(f"2025 Company ATV: ₹{company_atv_2025:,.2f}")
print(f"2025 Company UPT: {company_upt_2025:.2f}")

2025 Company Margin: 33.73%
2025 Company ATV: ₹1,475.94
2025 Company UPT: 1.87


In [468]:
conditions = [
    (
        (cross_store_analysis["Target_Achievement_Pct"] >= 100)
        & (cross_store_analysis["Gross_Margin_Pct"] >= company_margin_2025)
    ),
    (
        (cross_store_analysis["Target_Achievement_Pct"] >= 100)
        & (cross_store_analysis["Gross_Margin_Pct"] < company_margin_2025)
    ),
    (
        (cross_store_analysis["Target_Achievement_Pct"] < 100)
        & (cross_store_analysis["Gross_Margin_Pct"] >= company_margin_2025)
    )
]

choices = [
    "Strong Performer",
    "Margin Risk",
    "Growth Opportunity"
]

cross_store_analysis["Performance_Segment"] = np.select(
    conditions,
    choices,
    default="Target & Margin Risk"
)

In [469]:
cross_store_analysis[[
    "Store_ID",
    "Store_Name",
    "Net_Sales",
    "Gross_Margin_Pct",
    "ATV",
    "UPT",
    "Target_Achievement_Pct",
    "Months_Achieved",
    "Performance_Segment"
]].sort_values(
    ["Performance_Segment", "Target_Achievement_Pct"],
    ascending=[True, False]
)

,Store_ID,Store_Name,Net_Sales,Gross_Margin_Pct,ATV,UPT,Target_Achievement_Pct,Months_Achieved,Performance_Segment
19,ST020,Kolkata Store 2,490558.25,35.04,1587.57,1.97,98.20,9,Growth Opportunity
10,ST011,Ahmedabad Store 1,454261.30,35.01,1504.18,1.86,96.75,5,Growth Opportunity
16,ST017,Chennai Store 1,466128.00,34.97,1528.29,1.91,96.54,3,Growth Opportunity
7,ST008,Mumbai Store 2,439742.75,34.21,1437.07,1.85,95.69,8,Growth Opportunity
4,ST005,Jaipur Store 1,489522.15,34.98,1515.55,1.90,94.75,3,Growth Opportunity
6,ST007,Mumbai Store 1,461595.35,34.60,1503.57,1.89,92.89,4,Growth Opportunity
15,ST016,Hyderabad Store 2,475113.70,34.25,1369.20,1.74,91.63,4,Growth Opportunity
11,ST012,Ahmedabad Store 2,382969.85,34.29,1353.25,1.73,87.53,4,Growth Opportunity
22,ST023,Patna Store 1,440624.45,34.07,1463.87,1.85,84.17,2,Growth Opportunity
2,ST003,Lucknow Store 1,505498.60,33.11,1500.00,1.90,117.50,8,Margin Risk
